// @bitgo/sdk-core/dist/src/bitgo/coinFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalCoinFactory = exports.CoinFactory = void 0;
/**
 * @prettier
 */
const statics_1 = require("@bitgo/statics");
const errors_1 = require("./errors");
class CoinFactory {
    constructor() {
        this.coinConstructors = new Map();
    }
    /**
     * @param name Name of coin or address
     * @returns {(Object|undefined)}
     */
    getCoinConstructor(name) {
        if (this.coinConstructors === undefined) {
            this.coinConstructors = new Map();
        }
        return this.coinConstructors.get(name);
    }
    /**
     * @param name Name of coin or address
     * @param coin Coin plugin's constructor
     * @throws Error
     */
    register(name, coin) {
        this.coinConstructors.set(name, coin);
    }
    /**
     * @param bitgo Instance of BitGo
     * @param name Name of coin or address
     * @throws CoinNotDefinedError
     * @throws UnsupportedCoinError
     */
    getInstance(bitgo, name) {
        let staticsCoin;
        try {
            staticsCoin = statics_1.coins.get(name);
        }
        catch (e) {
            if (!(e instanceof statics_1.CoinNotDefinedError)) {
                throw e;
            }
        }
        const constructor = this.getCoinConstructor(name);
        if (constructor) {
            return constructor(bitgo, staticsCoin);
        }
        throw new errors_1.UnsupportedCoinError(name);
    }
}
exports.CoinFactory = CoinFactory;
exports.GlobalCoinFactory = new CoinFactory();
