// @bitgo/sdk-coin-ltc/dist/src/tltc.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tltc = void 0;
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const ltc_1 = require("./ltc");
class Tltc extends ltc_1.Ltc {
    constructor(bitgo) {
        super(bitgo, utxolib.networks.litecoinTest);
        this.altScriptHash = utxolib.networks.testnet.scriptHash;
        // support alt destinations on test
        this.supportAltScriptDestination = false;
    }
    static createInstance(bitgo) {
        return new Tltc(bitgo);
    }
    getChain() {
        return 'tltc';
    }
    getFullName() {
        return 'Testnet Litecoin';
    }
}
exports.Tltc = Tltc;
