// @bitgo/sdk-coin-near/dist/src/lib/contractCallWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractCallWrapper = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const sdk_core_1 = require("@bitgo/sdk-core");
/**
 * Contains parameters to call a Near Function Call action
 */
class ContractCallWrapper {
    /** Set method contract name */
    set methodName(methodName) {
        this._methodName = methodName;
    }
    /** Get method contract name */
    get methodName() {
        return this._methodName;
    }
    /** Set gas, expresed on yocto */
    set gas(gas) {
        if (!this.isValidAmount(new bignumber_js_1.default(gas))) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid gas value');
        }
        this._gas = gas;
    }
    /** Get gas, expresed on yocto*/
    get gas() {
        return this._gas;
    }
    /** Set deposit, expresed on yocto */
    set deposit(deposit) {
        if (!this.isValidAmount(new bignumber_js_1.default(deposit))) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid deposit value');
        }
        this._deposit = deposit;
    }
    /** Get deposit, expresed on yocto */
    get deposit() {
        return this._deposit;
    }
    /** Get args, which are the parameters of a method */
    set args(args) {
        this._args = args;
    }
    /** Set args, which are the parameters of a method */
    get args() {
        return this._args;
    }
    /**
     * Get all parameters of the contractCallWrapper
     * @returns {FunctionCall}
     */
    getParams() {
        return {
            methodName: this._methodName,
            args: this._args,
            gas: this._gas,
            deposit: this._deposit,
        };
    }
    isValidAmount(value) {
        return !value.isLessThan(0);
    }
}
exports.ContractCallWrapper = ContractCallWrapper;
