// @bitgo/sdk-core/dist/src/bitgo/webhook/webhooks.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = void 0;
/**
 * User webhook handler for v2 coins and tokens
 *
 * @prettier
 */
const _ = __importStar(require("lodash"));
const common = __importStar(require("../../common"));
class Webhooks {
    constructor(bitgo, baseCoin) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
    }
    /**
     * Fetch list of user webhooks
     *
     * @returns {*}
     */
    async list() {
        return await this.bitgo.get(this.baseCoin.url('/webhooks')).result();
    }
    /**
     * Add new user webhook
     *
     * @param params
     * @returns {*}
     */
    async add({ url, type, label, numConfirmations = 0 }) {
        common.validateParams({ url, type, label, numConfirmations }, ['url', 'type'], ['string', 'numConfirmations']);
        return await this.bitgo.post(this.baseCoin.url('/webhooks')).send({ url, type, label, numConfirmations }).result();
    }
    /**
     * Remove user webhook
     *
     * @param params
     * @returns {*}
     */
    async remove(params) {
        common.validateParams(params, ['url', 'type'], []);
        return this.bitgo.del(this.baseCoin.url('/webhooks')).send(params).result();
    }
    /**
     * Fetch list of webhook notifications for the user
     *
     * @param params
     * @returns {*}
     */
    async listNotifications(params = {}) {
        const queryProperties = [];
        if (params.prevId) {
            if (!_.isString(params.prevId)) {
                throw new Error('invalid prevId argument, expecting string');
            }
            queryProperties.push('prevId');
        }
        if (params.limit) {
            if (!_.isNumber(params.limit)) {
                throw new Error('invalid limit argument, expecting number');
            }
            queryProperties.push('limit');
        }
        const query = _.pick(params, queryProperties);
        return this.bitgo.get(this.baseCoin.url('/webhooks/notifications')).query(query).result();
    }
    /**
     * Simulate a user webhook
     *
     * @param params
     * @returns {*}
     */
    async simulate(params) {
        common.validateParams(params, ['webhookId', 'blockId'], []);
        const webhookId = params.webhookId;
        return this.bitgo
            .post(this.baseCoin.url('/webhooks/' + webhookId + '/simulate'))
            .send(params)
            .result();
    }
}
exports.Webhooks = Webhooks;
