// @bitgo/sdk-core/dist/src/account-lib/staking/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOperationConfig = exports.WithdrawMethodId = exports.ActivateMethodId = exports.UnvoteMethodId = exports.VoteMethodId = exports.UnlockMethodId = exports.LockMethodId = void 0;
const statics_1 = require("@bitgo/statics");
const baseCoin_1 = require("../baseCoin");
exports.LockMethodId = '0xf83d08ba'; // lock()
exports.UnlockMethodId = '0x6198e339'; // unlock()
exports.VoteMethodId = '0x580d747a'; // vote()
exports.UnvoteMethodId = '0x6e198475'; // revokeActive()
exports.ActivateMethodId = '0x1c5a9d9c'; // activate()
exports.WithdrawMethodId = '0x2e1a7d4d'; // withdraw()
const LockedGoldTestnetAddress = '0x94c3e6675015d8479b648657e7ddfcd938489d0d';
const LockedGoldMainnetAddress = '0x6cc083aed9e3ebe302a6336dbc7c921c9f03349e';
const ElectionTestnetAddress = '0x11fe523f93cac185d12cb39cc3bd279d2de524f8';
const ElectionMainnetAddress = '0x8d6677192144292870907e3fa8a5527fe55a7ff6';
const operations = {
    [baseCoin_1.StakingOperationTypes.LOCK]: {
        [statics_1.NetworkType.TESTNET]: {
            contractAddress: LockedGoldTestnetAddress,
            methodId: exports.LockMethodId,
            types: [],
        },
        [statics_1.NetworkType.MAINNET]: {
            contractAddress: LockedGoldMainnetAddress,
            methodId: exports.LockMethodId,
            types: [],
        },
    },
    [baseCoin_1.StakingOperationTypes.UNLOCK]: {
        [statics_1.NetworkType.TESTNET]: {
            contractAddress: LockedGoldTestnetAddress,
            methodId: exports.UnlockMethodId,
            types: ['uint256'],
        },
        [statics_1.NetworkType.MAINNET]: {
            contractAddress: LockedGoldMainnetAddress,
            methodId: exports.UnlockMethodId,
            types: ['uint256'],
        },
    },
    [baseCoin_1.StakingOperationTypes.VOTE]: {
        [statics_1.NetworkType.TESTNET]: {
            contractAddress: ElectionTestnetAddress,
            methodId: exports.VoteMethodId,
            types: ['address', 'uint256', 'address', 'address'],
        },
        [statics_1.NetworkType.MAINNET]: {
            contractAddress: ElectionMainnetAddress,
            methodId: exports.VoteMethodId,
            types: ['address', 'uint256', 'address', 'address'],
        },
    },
    [baseCoin_1.StakingOperationTypes.UNVOTE]: {
        [statics_1.NetworkType.TESTNET]: {
            contractAddress: ElectionTestnetAddress,
            methodId: exports.UnvoteMethodId,
            types: ['address', 'uint256', 'address', 'address', 'uint256'],
        },
        [statics_1.NetworkType.MAINNET]: {
            contractAddress: ElectionMainnetAddress,
            methodId: exports.UnvoteMethodId,
            types: ['address', 'uint256', 'address', 'address', 'uint256'],
        },
    },
    [baseCoin_1.StakingOperationTypes.ACTIVATE]: {
        [statics_1.NetworkType.TESTNET]: {
            contractAddress: ElectionTestnetAddress,
            methodId: exports.ActivateMethodId,
            types: ['address'],
        },
        [statics_1.NetworkType.MAINNET]: {
            contractAddress: ElectionMainnetAddress,
            methodId: exports.ActivateMethodId,
            types: ['address'],
        },
    },
    [baseCoin_1.StakingOperationTypes.WITHDRAW]: {
        [statics_1.NetworkType.TESTNET]: {
            contractAddress: LockedGoldTestnetAddress,
            methodId: exports.WithdrawMethodId,
            types: ['uint256'],
        },
        [statics_1.NetworkType.MAINNET]: {
            contractAddress: LockedGoldMainnetAddress,
            methodId: exports.WithdrawMethodId,
            types: ['uint256'],
        },
    },
};
/**
 * @param type
 * @param network
 */
function getOperationConfig(type, network) {
    return operations[type][network];
}
exports.getOperationConfig = getOperationConfig;
