// @bitgo/public-types/src/schema/webhook/webhookType.ts
import * as t from "io-ts";

export enum WebhookTypeEnum {
  // Webhook for txRequest.state changes
  "txRequest" = "txRequest",
  // Webhook for any txRequest.transactions[].state changes
  "txRequestTransaction" = "txRequestTransaction",
  "transfer" = "transfer",
  "transaction" = "transaction",
  "transactionRemoved" = "transactionRemoved",
  "transactionExpire" = "transactionExpire",
  "pendingapproval" = "pendingapproval",
  "block" = "block",
  "admin" = "admin",
  "address_confirmation" = "address_confirmation",
  "lowFee" = "lowFee",

  // Specific to User Webhooks
  "wallet_confirmation" = "wallet_confirmation",

  "wireWithdrawal" = "wireWithdrawal",
  "bankAccount" = "bankAccount",
}

export const WebhookType = t.keyof(WebhookTypeEnum, "WebhookType");
export type WebhookType = t.TypeOf<typeof WebhookType>;

export enum WalletWebhookTypeEnum {
  "txRequest" = "txRequest",
  "txRequestTransaction" = "txRequestTransaction",
  "transfer" = "transfer",
  "transaction" = "transaction",
  "transactionRemoved" = "transactionRemoved",
  "transactionExpire" = "transactionExpire",
  "pendingapproval" = "pendingapproval",
  "block" = "block",
  "admin" = "admin",
  "address_confirmation" = "address_confirmation",
  "lowFee" = "lowFee",
  "wireWithdrawal" = "wireWithdrawal",
}

export const WalletWebhookType = t.keyof(
  WalletWebhookTypeEnum,
  "WalletWebhookType",
);
export type WalletWebhookType = t.TypeOf<typeof WalletWebhookType>;

export enum EnterpriseWebhookTypeEnum {
  "bankAccount" = "bankAccount",
}
export const EnterpriseWebhookType = t.keyof(
  EnterpriseWebhookTypeEnum,
  "EnterpriseWebhookType",
);
export type EnterpriseWebhookType = t.TypeOf<typeof EnterpriseWebhookType>;
