// @bitgo/utxo-lib/dist/src/address.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.toBech32 = exports.fromBech32 = exports.fromBase58Check = exports.toBase58Check = exports.toOutputScript = exports.fromOutputScript = void 0;
const bitcoinjs = require("bitcoinjs-lib");
const zcashAddress = require("../src/bitgo/zcash/address");
const networks_1 = require("./networks");
const index_1 = require("./index");
function fromOutputScript(outputScript, network) {
    if ((0, networks_1.isValidNetwork)(network) && (0, networks_1.isZcash)(network)) {
        return zcashAddress.fromOutputScript(outputScript, network);
    }
    // We added p2tr payments from our forked bitcoinjs-lib to utxo-lib instead. Our bitcoinjs fork will no longer have
    // p2tr support so utxo-lib should take care of retrieving a p2tr address from outputScript and bitcoinjs-lib can
    // handle the other type of payments.
    try {
        return index_1.p2trPayments.p2tr({ output: outputScript, network }).address;
    }
    catch (e) {
        // noop. try the bitcoinjs method
    }
    return bitcoinjs.address.fromOutputScript(outputScript, network);
}
exports.fromOutputScript = fromOutputScript;
function toOutputScript(address, network) {
    if ((0, networks_1.isValidNetwork)(network) && (0, networks_1.isZcash)(network)) {
        return zcashAddress.toOutputScript(address, network);
    }
    return bitcoinjs.address.toOutputScript(address, network);
}
exports.toOutputScript = toOutputScript;
function toBase58Check(hash, version, network) {
    if ((0, networks_1.isValidNetwork)(network) && (0, networks_1.isZcash)(network)) {
        return zcashAddress.toBase58Check(hash, version);
    }
    return bitcoinjs.address.toBase58Check(hash, version);
}
exports.toBase58Check = toBase58Check;
function fromBase58Check(address, network) {
    if ((0, networks_1.isValidNetwork)(network) && (0, networks_1.isZcash)(network)) {
        return zcashAddress.fromBase58Check(address);
    }
    return bitcoinjs.address.fromBase58Check(address);
}
exports.fromBase58Check = fromBase58Check;
_a = bitcoinjs.address, exports.fromBech32 = _a.fromBech32, exports.toBech32 = _a.toBech32;
