// @bitgo/abstract-eth/dist/src/lib/transferBuilders/transferBuilderERC721.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC721TransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("ethers/lib/utils");
const contractCall_1 = require("../contractCall");
const utils_2 = require("../utils");
const baseNFTTransferBuilder_1 = require("./baseNFTTransferBuilder");
const walletUtil_1 = require("../walletUtil");
const statics_1 = require("@bitgo/statics");
class ERC721TransferBuilder extends baseNFTTransferBuilder_1.BaseNFTTransferBuilder {
    constructor(serializedData) {
        super(serializedData);
        this.bytes(0);
        if (serializedData) {
            this.decodeTransferData(serializedData);
        }
    }
    coin(coin) {
        this._coin = statics_1.coins.get(coin);
        this._nativeCoinOperationHashPrefix = this._coin.network.nativeCoinOperationHashPrefix;
        return this;
    }
    tokenContractAddress(address) {
        if ((0, utils_2.isValidEthAddress)(address)) {
            this._tokenContractAddress = address;
            return this;
        }
        throw new sdk_core_1.InvalidParameterValueError('Invalid address');
    }
    tokenId(token) {
        this._tokenId = token;
        return this;
    }
    bytes(bytesInNumber) {
        this._bytes = (0, utils_1.hexZeroPad)((0, utils_1.hexlify)(bytesInNumber), 32);
        return this;
    }
    build() {
        const types = walletUtil_1.ERC721SafeTransferTypes;
        const values = [this._fromAddress, this._toAddress, this._tokenId, this._bytes];
        const contractCall = new contractCall_1.ContractCall(walletUtil_1.ERC721SafeTransferTypeMethodId, types, values);
        return contractCall.serialize();
    }
    signAndBuild(chainId) {
        this._chainId = chainId;
        if (this.hasMandatoryFields()) {
            this._data = this.build();
            return (0, utils_2.sendMultiSigData)(this._tokenContractAddress, // to
            '0', // dummy amount value
            this._data, this._expirationTime, this._sequenceId, this.getSignature());
        }
        throw new sdk_core_1.BuildTransactionError(`Missing transfer mandatory fields. 
       Destination (to) address, Source (from) address, sequenceID, the token contract address and tokenID are mandatory`);
    }
    hasMandatoryFields() {
        return (this._tokenId !== undefined &&
            this._toAddress !== undefined &&
            this._fromAddress !== undefined &&
            this._tokenContractAddress !== undefined &&
            this._sequenceId !== undefined);
    }
    decodeTransferData(data) {
        const transferData = (0, utils_2.decodeERC721TransferData)(data);
        this._toAddress = transferData.to;
        this._fromAddress = transferData.from;
        this._expirationTime = transferData.expireTime;
        this._sequenceId = transferData.sequenceId;
        this._signature = transferData.signature;
        this._tokenContractAddress = transferData.tokenContractAddress;
        this._tokenId = transferData.tokenId;
        if (transferData.data) {
            this._data = transferData.data;
        }
    }
}
exports.ERC721TransferBuilder = ERC721TransferBuilder;
