// @bitgo/sdk-core/dist/src/account-lib/mpc/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPaillierPublicKey = exports.clamp = exports.bigIntToBufferBE = exports.bigIntFromBufferBE = exports.bigIntFromU8ABE = exports.bigIntToBufferLE = exports.bigIntFromBufferLE = exports.combineRound4DklsDsgMessages = void 0;
const sdk_lib_mpc_1 = require("@bitgo/sdk-lib-mpc");
Object.defineProperty(exports, "bigIntFromBufferLE", { enumerable: true, get: function () { return sdk_lib_mpc_1.bigIntFromBufferLE; } });
Object.defineProperty(exports, "bigIntToBufferLE", { enumerable: true, get: function () { return sdk_lib_mpc_1.bigIntToBufferLE; } });
Object.defineProperty(exports, "bigIntFromU8ABE", { enumerable: true, get: function () { return sdk_lib_mpc_1.bigIntFromU8ABE; } });
Object.defineProperty(exports, "bigIntFromBufferBE", { enumerable: true, get: function () { return sdk_lib_mpc_1.bigIntFromBufferBE; } });
Object.defineProperty(exports, "bigIntToBufferBE", { enumerable: true, get: function () { return sdk_lib_mpc_1.bigIntToBufferBE; } });
Object.defineProperty(exports, "clamp", { enumerable: true, get: function () { return sdk_lib_mpc_1.clamp; } });
Object.defineProperty(exports, "getPaillierPublicKey", { enumerable: true, get: function () { return sdk_lib_mpc_1.getPaillierPublicKey; } });
/**
 * Combines serialized partial signatures from parties participating in DSG.
 * @param round4DsgMessages - round 4 serialized broadcast messages payloads from participating parties
 * @returns {DklsTypes.SerializedDklsSignature} - serialized final signature
 */
function combineRound4DklsDsgMessages(round4DsgMessages) {
    var _a;
    const round4DsgMessagesDeser = round4DsgMessages.map(sdk_lib_mpc_1.DklsTypes.deserializeBroadcastMessage);
    const signatureR = (_a = round4DsgMessagesDeser.find((m) => m.signatureR !== undefined)) === null || _a === void 0 ? void 0 : _a.signatureR;
    if (!signatureR) {
        throw Error('None of the round 4 Dkls messages contain a Signature.R value.');
    }
    const signatureDeser = sdk_lib_mpc_1.DklsUtils.combinePartialSignatures(round4DsgMessagesDeser.map((m) => m.payload), Buffer.from(signatureR).toString('hex'));
    return {
        R: Buffer.from(signatureDeser.R).toString('hex'),
        S: Buffer.from(signatureDeser.S).toString('hex'),
    };
}
exports.combineRound4DklsDsgMessages = combineRound4DklsDsgMessages;
