// @bitgo/sdk-coin-trx/dist/src/trxToken.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrxToken = void 0;
const trx_1 = require("./trx");
const statics_1 = require("@bitgo/statics");
const builder_1 = require("./lib/builder");
const assert_1 = __importDefault(require("assert"));
class TrxToken extends trx_1.Trx {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('trx') : statics_1.coins.get('ttrx');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new TrxToken(bitgo, config);
    }
    static createTokenConstructors() {
        const tokensCtors = [];
        for (const token of [...statics_1.tokens.bitcoin.trx.tokens, ...statics_1.tokens.testnet.trx.tokens]) {
            const tokenConstructor = TrxToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get type() {
        return this.tokenConfig.type;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get tokenContractAddress() {
        return this.tokenConfig.tokenContractAddress;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Tron Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return false;
    }
    /**
     * Flag for sending data along with transactions
     * @returns {boolean} True if okay to send tx data (ETH), false otherwise
     */
    transactionDataAllowed() {
        return false;
    }
    async verifyTransaction(params) {
        const { txPrebuild: txPrebuild, txParams: txParams } = params;
        (0, assert_1.default)(txPrebuild.txHex, new Error('missing required tx prebuild property txHex'));
        const rawTx = txPrebuild.txHex;
        const txBuilder = (0, builder_1.getBuilder)(this.getChain()).from(rawTx);
        const tx = await txBuilder.build();
        const recipients = txParams.recipients || txPrebuild.txInfo.recipients;
        if (!recipients) {
            throw new Error('missing required property recipients');
        }
        if (recipients[0].address === tx.outputs[0].address && recipients[0].amount === tx.outputs[0].value) {
            return true;
        }
        else {
            throw new Error('Tx outputs does not match with expected txParams recipients');
        }
    }
}
exports.TrxToken = TrxToken;
