// @bitgo/sdk-coin-avaxp/dist/test/unit/lib/importC2PTxBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
require("should");
const errorMessage = __importStar(require("../../resources/errors"));
const lib_1 = require("../../../src/lib");
const statics_1 = require("@bitgo/statics");
const importC_1 = require("../../resources/tx/importC");
const signFlowTestSuit_1 = __importDefault(require("./signFlowTestSuit"));
describe('AvaxP Import C2P Tx Builder', () => {
    const factory = new lib_1.TransactionBuilderFactory(statics_1.coins.get('tavaxp'));
    describe('validate txBuilder fields', () => {
        const txBuilder = factory.getImportInCBuilder();
        it('should fail validate Utxos empty string', () => {
            assert_1.default.throws(() => {
                txBuilder.validateUtxos([]);
            }, (e) => e.message === errorMessage.ERROR_UTXOS_EMPTY);
        });
        it('should fail validate Utxos without amount field', () => {
            assert_1.default.throws(() => {
                txBuilder.validateUtxos([{ outputID: '' }]);
            }, (e) => e.message === errorMessage.ERROR_UTXOS_AMOUNT);
        });
    });
    (0, signFlowTestSuit_1.default)({
        transactionType: 'Import C2P',
        newTxFactory: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tavaxp')),
        newTxBuilder: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tavaxp'))
            .getImportInCBuilder()
            .threshold(importC_1.IMPORT_C.threshold)
            .fromPubKey(importC_1.IMPORT_C.pAddresses)
            .utxos(importC_1.IMPORT_C.outputs)
            .to(importC_1.IMPORT_C.to)
            .feeRate(importC_1.IMPORT_C.fee),
        unsignedTxHex: importC_1.IMPORT_C.unsignedTxHex,
        halfsigntxHex: importC_1.IMPORT_C.halfsigntxHex,
        fullsigntxHex: importC_1.IMPORT_C.fullsigntxHex,
        privKey: {
            prv1: importC_1.IMPORT_C.privKey.prv1,
            prv2: importC_1.IMPORT_C.privKey.prv2,
        },
    });
    (0, signFlowTestSuit_1.default)({
        transactionType: 'Import C2P recovery',
        newTxFactory: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tavaxp')),
        newTxBuilder: () => new lib_1.TransactionBuilderFactory(statics_1.coins.get('tavaxp'))
            .getImportInCBuilder()
            .threshold(importC_1.IMPORT_C.threshold)
            .fromPubKey(importC_1.IMPORT_C.pAddresses)
            .utxos(importC_1.IMPORT_C.outputs)
            .to(importC_1.IMPORT_C.to)
            .feeRate(importC_1.IMPORT_C.fee)
            .recoverMode(),
        unsignedTxHex: importC_1.IMPORT_C.rUnsignedTxHex,
        halfsigntxHex: importC_1.IMPORT_C.rHalfsigntxHex,
        fullsigntxHex: importC_1.IMPORT_C.rFullsigntxHex,
        privKey: {
            prv1: importC_1.IMPORT_C.privKey.prv3,
            prv2: importC_1.IMPORT_C.privKey.prv2,
        },
    });
});
