// @bitgo/abstract-eth/dist/src/lib/walletUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createV4ForwarderTypes = exports.createV1ForwarderTypes = exports.ERC1155BatchTransferTypes = exports.ERC1155SafeTransferTypes = exports.ERC721SafeTransferTypes = exports.sendMultiSigTokenTypes = exports.sendMultiSigTypes = exports.flushCoinsTypes = exports.flushTokensTypesv4 = exports.flushTokensTypes = exports.createV1WalletTypes = exports.walletSimpleConstructor = exports.defaultWalletVersion = exports.defaultForwarderVersion = exports.ERC1155BatchTransferTypeMethodId = exports.ERC1155SafeTransferTypeMethodId = exports.ERC721SafeTransferTypeMethodId = exports.flushCoinsMethodId = exports.flushForwarderTokensMethodIdV4 = exports.flushForwarderTokensMethodId = exports.recoveryWalletInitializationFirstBytes = exports.walletInitializationFirstBytes = exports.createForwarderMethodId = exports.v1CreateWalletMethodId = exports.v1WalletInitializationFirstBytes = exports.v4CreateForwarderMethodId = exports.v1CreateForwarderMethodId = exports.sendMultisigTokenMethodId = exports.sendMultisigMethodId = void 0;
exports.sendMultisigMethodId = '0x39125215';
exports.sendMultisigTokenMethodId = '0x0dcd7a6c';
exports.v1CreateForwarderMethodId = '0xfb90b320';
exports.v4CreateForwarderMethodId = '0x13b2f75c';
exports.v1WalletInitializationFirstBytes = '0x60806040';
exports.v1CreateWalletMethodId = '0x7117f3fa';
exports.createForwarderMethodId = '0xa68a76cc';
exports.walletInitializationFirstBytes = '0x60606040';
exports.recoveryWalletInitializationFirstBytes = '0x60c06040';
exports.flushForwarderTokensMethodId = '0x2da03409';
exports.flushForwarderTokensMethodIdV4 = '0x3ef13367';
exports.flushCoinsMethodId = '0x6b9f96ea';
exports.ERC721SafeTransferTypeMethodId = '0xb88d4fde';
exports.ERC1155SafeTransferTypeMethodId = '0xf242432a';
exports.ERC1155BatchTransferTypeMethodId = '0x2eb2c2d6';
exports.defaultForwarderVersion = 0;
exports.defaultWalletVersion = 0;
exports.walletSimpleConstructor = ['address[]'];
exports.createV1WalletTypes = ['address[]', 'bytes32'];
exports.flushTokensTypes = ['address', 'address'];
exports.flushTokensTypesv4 = ['address'];
exports.flushCoinsTypes = [];
exports.sendMultiSigTypes = ['address', 'uint', 'bytes', 'uint', 'uint', 'bytes'];
exports.sendMultiSigTokenTypes = ['address', 'uint', 'address', 'uint', 'uint', 'bytes'];
exports.ERC721SafeTransferTypes = ['address', 'address', 'uint256', 'bytes'];
exports.ERC1155SafeTransferTypes = ['address', 'address', 'uint256', 'uint256', 'bytes'];
exports.ERC1155BatchTransferTypes = ['address', 'address', 'uint256[]', 'uint256[]', 'bytes'];
exports.createV1ForwarderTypes = ['address', 'bytes32'];
exports.createV4ForwarderTypes = ['address', 'address', 'bytes32'];
