// @bitgo/sdk-coin-algo/dist/src/algoToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlgoToken = void 0;
/**
 * @prettier
 */
const algo_1 = require("./algo");
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
class AlgoToken extends algo_1.Algo {
    constructor(bitgo, tokenConfig) {
        super(bitgo);
        this.tokenConfig = tokenConfig;
        const match = this.tokenConfig.type.match(AlgoToken.tokenNamePattern) || [];
        const tokenCoin = match[1];
        this._code = match[2];
        const token = match[3];
        if (tokenCoin !== tokenConfig.coin) {
            throw new sdk_core_1.BitGoJsError(`invalid coin found in token: ${this.tokenConfig.type}`);
        }
        if (!token) {
            throw new sdk_core_1.BitGoJsError(`invalid token: ${this.tokenConfig.type}`);
        }
    }
    static createTokenConstructor(config) {
        return (bitgo) => new AlgoToken(bitgo, config);
    }
    static createTokenConstructors() {
        const tokensCtors = [];
        statics_1.formattedAlgoTokens.forEach((config) => {
            tokensCtors.push({ name: config.type, coinConstructor: AlgoToken.createTokenConstructor(config) });
            if (config.alias) {
                tokensCtors.push({ name: config.alias, coinConstructor: AlgoToken.createTokenConstructor(config) });
            }
        });
        return tokensCtors;
    }
    get type() {
        return this.tokenConfig.type;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get code() {
        return this._code;
    }
    get issuer() {
        return undefined; // Not defined for Algorand
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Algo Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    valuelessTransferAllowed() {
        return true;
    }
}
exports.AlgoToken = AlgoToken;
AlgoToken.tokenNamePattern = /^([^:]+):(?:([^.]+)-)?([0-9]+)$/;
AlgoToken.tokenPattern = /[0-9]/;
