// @bitgo/sdk-coin-dot/dist/src/lib/iface_utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAddress = exports.isStakeBatchCallPayeeAccount = exports.isStakeBatchCallPayeeController = exports.isStakeBatchCallPayeeStash = exports.isStakeBatchCallPayeeStaked = exports.getDelegateAddress = exports.isAccountId = void 0;
/**
 * Returns true if value is of type AccountId, false otherwise.
 *
 * @param value The object to test.
 *
 * @return true if value is of type AccountId, false otherwise.
 */
function isAccountId(value) {
    return value.hasOwnProperty('id');
}
exports.isAccountId = isAccountId;
/**
 * Extracts the proxy address being added from an add proxy batch call or an add proxy call.

 * @param call A batched add proxy call or an add proxy call from which to extract the proxy
 * address.
 *
 * @return the proxy address being added from an add proxy batch call or an add proxy call.
 */
function getDelegateAddress(call) {
    if (isAccountId(call.delegate)) {
        return call.delegate.id;
    }
    else {
        return call.delegate;
    }
}
exports.getDelegateAddress = getDelegateAddress;
/**
 * Returns true if value is of type StakeBatchCallPayeeStaked, false otherwise.
 *
 * @param value The object to test.
 *
 * @return true if value is of type StakeBatchCallPayeeStaked, false otherwise.
 */
function isStakeBatchCallPayeeStaked(value) {
    return value.hasOwnProperty('staked');
}
exports.isStakeBatchCallPayeeStaked = isStakeBatchCallPayeeStaked;
/**
 * Returns true if value is of type StakeBatchCallPayeeStash, false otherwise.
 *
 * @param value The object to test.
 *
 * @return true if value is of type StakeBatchCallPayeeStash, false otherwise.
 */
function isStakeBatchCallPayeeStash(value) {
    return value.hasOwnProperty('stash');
}
exports.isStakeBatchCallPayeeStash = isStakeBatchCallPayeeStash;
/**
 * Returns true if value is of type StakeBatchCallPayeeController, false otherwise.
 *
 * @param value The object to test.
 *
 * @return true if value is of type StakeBatchCallPayeeController, false otherwise.
 */
function isStakeBatchCallPayeeController(value) {
    return value.hasOwnProperty('controller');
}
exports.isStakeBatchCallPayeeController = isStakeBatchCallPayeeController;
/**
 * Returns true if value is of type StakeBatchCallPayeeAccount, false otherwise.
 *
 * @param value The object to test.
 *
 * @return true if value is of type StakeBatchCallPayeeAccount, false otherwise.
 */
function isStakeBatchCallPayeeAccount(value) {
    return (value.account !== undefined &&
        value.account !== null);
}
exports.isStakeBatchCallPayeeAccount = isStakeBatchCallPayeeAccount;
/**
 * Extracts the proxy address being added from ProxyArgs.

 * @param args the ProxyArgs object from which to extract the proxy address.
 *
 * @return the proxy address being added.
 */
function getAddress(args) {
    if (isAccountId(args.real)) {
        return args.real.id;
    }
    else {
        return args.real;
    }
}
exports.getAddress = getAddress;
