// @bitgo/sdk-coin-dot/dist/src/lib/keyPair.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair = void 0;
const keyring_1 = require("@polkadot/keyring");
const pair_1 = require("@polkadot/keyring/pair");
const sdk_core_1 = require("@bitgo/sdk-core");
const bs58_1 = __importDefault(require("bs58"));
const utils_1 = __importDefault(require("./utils"));
const TYPE = 'ed25519';
const keyring = new keyring_1.Keyring({ type: TYPE });
class KeyPair extends sdk_core_1.Ed25519KeyPair {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
     */
    constructor(source) {
        super(source);
    }
    /**
     * Helper function to create the KeyringPair for signing a dot transaction.
     *
     * @returns {KeyringPair} dot KeyringPair
     *
     * @see https://polkadot.js.org/docs/api/start/keyring
     */
    createPolkadotPair() {
        const secretKey = this.keyPair.prv ? new Uint8Array(Buffer.from(this.keyPair.prv, 'hex')) : undefined;
        const publicKey = new Uint8Array(Buffer.from(this.keyPair.pub, 'hex'));
        return (0, pair_1.createPair)({ toSS58: keyring.encodeAddress, type: TYPE }, { secretKey, publicKey });
    }
    /**
     // https://wiki.polkadot.network/docs/learn-accounts#address-format
     * Returns the address in either mainnet polkadot format (starts with 1)
     * or substrate format used for westend (starts with 5)
     */
    getAddress(format) {
        let encodedAddress = this.createPolkadotPair().address;
        encodedAddress = keyring.encodeAddress(encodedAddress, format);
        return encodedAddress;
    }
    /** @inheritdoc */
    getKeys() {
        const result = { pub: this.keyPair.pub };
        if (this.keyPair.prv) {
            result.prv = this.keyPair.prv;
        }
        return result;
    }
    /** @inheritdoc */
    recordKeysFromPrivateKeyInProtocolFormat(prv) {
        const decodedSeed = utils_1.default.decodeSeed(prv);
        const bufferFromSeed = Buffer.from(decodedSeed.seed);
        return utils_1.default.keyPairFromSeed(bufferFromSeed).keyPair;
    }
    /** @inheritdoc */
    recordKeysFromPublicKeyInProtocolFormat(pub) {
        const publicKey = keyring.addFromPair({
            // tss common pub is in base58 format and decodes to length of 32
            publicKey: (0, sdk_core_1.isBase58)(pub, 32) ? new Uint8Array(bs58_1.default.decode(pub)) : new Uint8Array(Buffer.from(pub, 'hex')),
            secretKey: new Uint8Array(),
        }).publicKey;
        return { pub: (0, sdk_core_1.toHex)(publicKey) };
    }
}
exports.KeyPair = KeyPair;
