// @bitgo/sdk-coin-dot/dist/src/lib/withdrawUnstakedBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WithdrawUnstakedBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const txwrapper_polkadot_1 = require("@substrate/txwrapper-polkadot");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const iface_1 = require("./iface");
const transactionBuilder_1 = require("./transactionBuilder");
const txnSchema_1 = require("./txnSchema");
class WithdrawUnstakedBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /**
     *
     * Frees up any unlocked chunks.
     * The balance can now be used by the stash account to do whatever it wants.
     *
     * @returns {UnsignedTransaction} an unsigned Dot transaction
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#withdrawunbondednum_slashing_spans-u32
     */
    buildTransaction() {
        const baseTxInfo = this.createBaseTxInfo();
        // TODO: The method changes the args type from number to string, verify it doesn't break anything
        return txwrapper_polkadot_1.methods.staking.withdrawUnbonded({
            numSlashingSpans: this._numSlashingSpans,
        }, baseTxInfo.baseTxInfo, baseTxInfo.options);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingWithdraw;
    }
    /**
     *
     * The number of slashing spans.
     *
     * @param {number} slashingSpans
     * @returns {WithdrawUnstakedBuilder} This withdrawUnstaked builder.
     *
     * @see https://polkadot.js.org/docs/substrate/extrinsics/#withdrawunbondednum_slashing_spans-u32
     */
    slashingSpans(slashingSpans) {
        this.validateValue(new bignumber_js_1.default(slashingSpans));
        this._numSlashingSpans = slashingSpans;
        return this;
    }
    /** @inheritdoc */
    validateDecodedTransaction(decodedTxn) {
        var _a;
        if (((_a = decodedTxn.method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.WithdrawUnbonded) {
            const txMethod = decodedTxn.method.args;
            const value = txMethod.numSlashingSpans;
            const validationResult = txnSchema_1.WithdrawUnstakedTransactionSchema.validate({ value });
            if (validationResult.error) {
                throw new sdk_core_1.InvalidTransactionError(`WithdrawUnstaked Transaction validation failed: ${validationResult.error.message}`);
            }
        }
    }
    /** @inheritdoc */
    fromImplementation(rawTransaction) {
        var _a, _b;
        const tx = super.fromImplementation(rawTransaction);
        if (((_a = this._method) === null || _a === void 0 ? void 0 : _a.name) === iface_1.MethodNames.WithdrawUnbonded) {
            const txMethod = this._method.args;
            this.slashingSpans(txMethod.numSlashingSpans);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${(_b = this._method) === null || _b === void 0 ? void 0 : _b.name}. Expected withdrawUnbonded`);
        }
        return tx;
    }
    /** @inheritdoc */
    validateTransaction(_) {
        super.validateTransaction(_);
        this.validateFields(this._numSlashingSpans);
    }
    validateFields(value) {
        const validationResult = txnSchema_1.WithdrawUnstakedTransactionSchema.validate({
            value,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`WithdrawUnstaked Builder Transaction validation failed: ${validationResult.error.message}`);
        }
    }
}
exports.WithdrawUnstakedBuilder = WithdrawUnstakedBuilder;
