// @bitgo/sdk-coin-avaxc/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getToken = exports.isValidEthPublicKey = exports.isValidEthPrivateKey = exports.isValidEthAddress = exports.getCommon = exports.sign = void 0;
const assert_1 = __importDefault(require("assert"));
const statics_1 = require("@bitgo/statics");
const ethereumjs_util_1 = require("ethereumjs-util");
const sdk_coin_eth_1 = require("@bitgo/sdk-coin-eth");
const sdk_core_1 = require("@bitgo/sdk-core");
const resources_1 = require("./resources");
/**
 * Signs the transaction using the appropriate algorithm
 *
 * @param {TxData} transactionData the transaction data to sign
 * @param {KeyPair} keyPair the signer's keypair
 * @returns {string} the transaction signed and encoded
 */
async function sign(transactionData, keyPair) {
    return sdk_coin_eth_1.Utils.signInternal(transactionData, keyPair, resources_1.testnetCommon);
}
exports.sign = sign;
const commons = new Map([
    [statics_1.NetworkType.MAINNET, resources_1.mainnetCommon],
    [statics_1.NetworkType.TESTNET, resources_1.testnetCommon],
]);
/**
 * @param {NetworkType} network either mainnet or testnet
 * @returns {EthereumCommon} Ethereum common configuration object
 */
function getCommon(network) {
    const common = commons.get(network);
    if (!common) {
        throw new sdk_core_1.InvalidTransactionError('Missing network common configuration');
    }
    return common;
}
exports.getCommon = getCommon;
/**
 * Returns whether or not the string is a valid C-Chain address in Eth format
 *
 * @param {string} address - the address string
 * @returns {boolean} - the validation result
 */
function isValidEthAddress(address) {
    return (0, ethereumjs_util_1.isValidAddress)(address);
}
exports.isValidEthAddress = isValidEthAddress;
/**
 * Returns whether or not the string is a valid C-Chain private key in Eth format
 *
 * @param {string} privateKey - the string formatted key
 * @returns {boolean} - the validation result
 */
function isValidEthPrivateKey(privateKey) {
    if (privateKey.length !== 64) {
        return false;
    }
    const privateKeyBuffer = Buffer.from(privateKey, 'hex');
    if (privateKeyBuffer.length !== 32) {
        return false;
    }
    return (0, ethereumjs_util_1.isValidPrivate)(privateKeyBuffer);
}
exports.isValidEthPrivateKey = isValidEthPrivateKey;
/**
 * Returns whether or not the string is a valid C-Chain public key in Eth format
 *
 * @param {string} publicKey - the uncompressed public key string
 * @returns {boolean} - the validation result
 */
function isValidEthPublicKey(publicKey) {
    // Uncompressed Eth Public Keys have a 04 prefix that needs to be removed in order to validate it.
    const publicKeyWithoutPrefix = publicKey.slice(2);
    const publicKeyBuffer = Buffer.from(publicKeyWithoutPrefix, 'hex');
    return (0, ethereumjs_util_1.isValidPublic)(publicKeyBuffer);
}
exports.isValidEthPublicKey = isValidEthPublicKey;
function getToken(tokenContractAddress, network) {
    const tokens = statics_1.coins.filter((coin) => {
        if (coin instanceof statics_1.AvaxERC20Token) {
            return (coin.network.type === network.type && coin.contractAddress.toLowerCase() === tokenContractAddress.toLowerCase());
        }
        return false;
    });
    // if length of tokens is 1, return the first, else return undefined
    // Can't directly index into tokens, or call `length`, so we use map to get an array
    const tokensArray = tokens.map((token) => token);
    if (tokensArray.length >= 1) {
        // there should never be two tokens with the same contract address, so we assert that here
        (0, assert_1.default)(tokensArray.length === 1, 'Found more than 1 token: ' + JSON.stringify(tokensArray));
        return tokensArray[0];
    }
    return undefined;
}
exports.getToken = getToken;
