// @bitgo/abstract-lightning/dist/src/abstractLightningCoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractLightningCoin = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
class AbstractLightningCoin extends sdk_core_1.BaseCoin {
    constructor(bitgo) {
        super(bitgo);
    }
    getBaseFactor() {
        throw new Error('Method not implemented.');
    }
    verifyTransaction(params) {
        throw new Error('Method not implemented.');
    }
    isWalletAddress(params) {
        throw new Error('Method not implemented.');
    }
    parseTransaction(params) {
        throw new Error('Method not implemented.');
    }
    generateKeyPair(seed) {
        throw new Error('Method not implemented.');
    }
    isValidPub(pub) {
        throw new Error('Method not implemented.');
    }
    isValidAddress(address) {
        throw new Error('Method not implemented.');
    }
    signTransaction(params) {
        throw new Error('Method not implemented.');
    }
}
exports.AbstractLightningCoin = AbstractLightningCoin;
