// @bitgo/sdk-core/dist/src/account-lib/baseCoin/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakingOperationTypes = exports.DotAddressFormat = exports.AddressFormat = exports.TransactionType = void 0;
/**
 * Internal metadata. Defines the type of transaction.
 */
var TransactionType;
(function (TransactionType) {
    TransactionType[TransactionType["Send"] = 0] = "Send";
    // Send ERC721 compliant tokens
    TransactionType[TransactionType["SendERC721"] = 1] = "SendERC721";
    // Send ERC1155 compliant tokens
    TransactionType[TransactionType["SendERC1155"] = 2] = "SendERC1155";
    // Initialize a wallet on-chain (e.g. Multi-sig contract deployment)
    TransactionType[TransactionType["WalletInitialization"] = 3] = "WalletInitialization";
    // Initialize an address on-chain(e.g. Forwarder contract deployment)
    TransactionType[TransactionType["AddressInitialization"] = 4] = "AddressInitialization";
    // Initialized an associated token account
    TransactionType[TransactionType["AssociatedTokenAccountInitialization"] = 5] = "AssociatedTokenAccountInitialization";
    // Flush tokens from a forwarder address to its base address
    TransactionType[TransactionType["FlushTokens"] = 6] = "FlushTokens";
    // Flush native coins (eg. ETH) from a forwarder address to base address
    TransactionType[TransactionType["FlushCoins"] = 7] = "FlushCoins";
    // Send a raw single-sig transaction
    TransactionType[TransactionType["SingleSigSend"] = 8] = "SingleSigSend";
    // Update an account on-chain (e.g. Public key revelation operation for Tezos)
    TransactionType[TransactionType["AccountUpdate"] = 9] = "AccountUpdate";
    // Lock
    TransactionType[TransactionType["StakingLock"] = 10] = "StakingLock";
    // Vote
    TransactionType[TransactionType["StakingVote"] = 11] = "StakingVote";
    // Unvote
    TransactionType[TransactionType["StakingUnvote"] = 12] = "StakingUnvote";
    // Activate
    TransactionType[TransactionType["StakingActivate"] = 13] = "StakingActivate";
    // Unlock
    TransactionType[TransactionType["StakingUnlock"] = 14] = "StakingUnlock";
    // Withdraw
    TransactionType[TransactionType["StakingWithdraw"] = 15] = "StakingWithdraw";
    // Handle smart contract calls
    TransactionType[TransactionType["ContractCall"] = 16] = "ContractCall";
    // Deactivate
    TransactionType[TransactionType["StakingDeactivate"] = 17] = "StakingDeactivate";
    // Batch of multiple transactions broadcast as a single transaction
    TransactionType[TransactionType["Batch"] = 18] = "Batch";
    // Claim the staking rewards
    TransactionType[TransactionType["StakingClaim"] = 19] = "StakingClaim";
    // Add delegator to p-chain
    // @deprecated
    TransactionType[TransactionType["AddDelegator"] = 20] = "AddDelegator";
    // Add validator to p-chain
    // @deprecated
    TransactionType[TransactionType["AddValidator"] = 21] = "AddValidator";
    // Cross chain transfer at Avax
    TransactionType[TransactionType["Export"] = 22] = "Export";
    TransactionType[TransactionType["Import"] = 23] = "Import";
    // Cross chain Recovery for evm like chains
    TransactionType[TransactionType["RecoveryWalletDeployment"] = 24] = "RecoveryWalletDeployment";
    // Add staking
    TransactionType[TransactionType["StakingAdd"] = 25] = "StakingAdd";
    // Staking pledge (e.g. ADA)
    TransactionType[TransactionType["StakingPledge"] = 26] = "StakingPledge";
    // Staking Authorize (e.g. SOL)
    TransactionType[TransactionType["StakingAuthorize"] = 27] = "StakingAuthorize";
    // Staking Authorize from raw message (e.g. SOL)
    TransactionType[TransactionType["StakingAuthorizeRaw"] = 28] = "StakingAuthorizeRaw";
    // Delegate
    TransactionType[TransactionType["StakingDelegate"] = 29] = "StakingDelegate";
    // Custom transaction (e.g. SUI)
    TransactionType[TransactionType["CustomTx"] = 30] = "CustomTx";
    TransactionType[TransactionType["StakingRedelegate"] = 31] = "StakingRedelegate";
    TransactionType[TransactionType["AddPermissionlessDelegator"] = 32] = "AddPermissionlessDelegator";
    TransactionType[TransactionType["AddPermissionlessValidator"] = 33] = "AddPermissionlessValidator";
    // Closing an associated token account (e.g. SOL)
    TransactionType[TransactionType["CloseAssociatedTokenAccount"] = 34] = "CloseAssociatedTokenAccount";
})(TransactionType = exports.TransactionType || (exports.TransactionType = {}));
/**
 * Generic list of encoding formats. Can be used as arguments for methods inputs.
 */
var AddressFormat;
(function (AddressFormat) {
    AddressFormat["hex"] = "hex";
    AddressFormat["base58"] = "base58";
    AddressFormat["mainnet"] = "mainnet";
    AddressFormat["testnet"] = "testnet";
})(AddressFormat = exports.AddressFormat || (exports.AddressFormat = {}));
// TODO(): create union type of all address formats enums
// list of ss58 format encoding types for Dot ecosystem
var DotAddressFormat;
(function (DotAddressFormat) {
    // format for westend (generic substrate) addresses
    DotAddressFormat[DotAddressFormat["substrate"] = 42] = "substrate";
    // format for polkadot mainnet addresses
    DotAddressFormat[DotAddressFormat["polkadot"] = 0] = "polkadot";
})(DotAddressFormat = exports.DotAddressFormat || (exports.DotAddressFormat = {}));
var StakingOperationTypes;
(function (StakingOperationTypes) {
    StakingOperationTypes[StakingOperationTypes["LOCK"] = 0] = "LOCK";
    StakingOperationTypes[StakingOperationTypes["VOTE"] = 1] = "VOTE";
    StakingOperationTypes[StakingOperationTypes["UNVOTE"] = 2] = "UNVOTE";
    StakingOperationTypes[StakingOperationTypes["ACTIVATE"] = 3] = "ACTIVATE";
    StakingOperationTypes[StakingOperationTypes["UNLOCK"] = 4] = "UNLOCK";
    StakingOperationTypes[StakingOperationTypes["WITHDRAW"] = 5] = "WITHDRAW";
})(StakingOperationTypes = exports.StakingOperationTypes || (exports.StakingOperationTypes = {}));
