// @bitgo/sdk-core/dist/src/account-lib/baseCoin/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DuplicateMethodError = exports.UnsupportedTokenError = exports.InvalidKey = exports.NotSupported = exports.NotImplementedError = exports.InvalidSignatureError = exports.InvalidParameterValueError = exports.ExtendTransactionError = exports.UtilsError = exports.BuildTransactionError = exports.SigningError = exports.ParseTransactionError = exports.ExtendableError = void 0;
/**
 * This base class ensures that our stack trace is captured properly but also that we have classes of errors
 * that can be found in a switch.
 */
class ExtendableError extends Error {
    constructor(message) {
        super(message);
        this.name = this.constructor.name;
        if (typeof Error.captureStackTrace === 'function') {
            Error.captureStackTrace(this, this.constructor);
        }
        else {
            this.stack = new Error(message).stack;
        }
    }
}
exports.ExtendableError = ExtendableError;
class ParseTransactionError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.ParseTransactionError = ParseTransactionError;
class SigningError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.SigningError = SigningError;
class BuildTransactionError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.BuildTransactionError = BuildTransactionError;
class UtilsError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.UtilsError = UtilsError;
class ExtendTransactionError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.ExtendTransactionError = ExtendTransactionError;
/**
 * Error for an invalid value for a contract method parameter
 */
class InvalidParameterValueError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.InvalidParameterValueError = InvalidParameterValueError;
/**
 * Error produced by not truthy signature values
 */
class InvalidSignatureError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.InvalidSignatureError = InvalidSignatureError;
/**
 * Error for a method that needs to be implemented
 */
class NotImplementedError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.NotImplementedError = NotImplementedError;
/**
 * Error for not supported features
 */
class NotSupported extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.NotSupported = NotSupported;
/**
 * Error for invalid seed, public, or private keys
 */
class InvalidKey extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.InvalidKey = InvalidKey;
/**
 * Error for invalid mint/token
 */
class UnsupportedTokenError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.UnsupportedTokenError = UnsupportedTokenError;
class DuplicateMethodError extends ExtendableError {
    constructor(message) {
        super(message);
    }
}
exports.DuplicateMethodError = DuplicateMethodError;
