// @bitgo/sdk-coin-hbar/dist/src/hbarToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HbarToken = void 0;
const statics_1 = require("@bitgo/statics");
const hbar_1 = require("./hbar");
class HbarToken extends hbar_1.Hbar {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('hbar') : statics_1.coins.get('thbar');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new HbarToken(bitgo, config);
    }
    static createTokenConstructors() {
        const tokensCtors = [];
        for (const token of [...statics_1.tokens.bitcoin.hbar.tokens, ...statics_1.tokens.testnet.hbar.tokens]) {
            const tokenConstructor = HbarToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return this.name;
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.HbarToken = HbarToken;
