// @bitgo/utxo-lib/dist/src/bitgo/legacysafe/index.js
"use strict";
/**
 * V1 Safe Wallets are the oldest type of wallets that BitGo supports. They were
 * created back in 2013-14 and don't use HD chains. Instead, they have only one
 * P2SH address per wallet whose redeem script uses uncompressed public keys.
 * */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLegacySafeOutputScript2of3 = exports.toCompressedPub = exports.toUncompressedPub = void 0;
const assert = require("assert");
const noble_ecc_1 = require("../../noble_ecc");
const networks_1 = require("../../networks");
const types_1 = require("../types");
const bitcoinjs = require("bitcoinjs-lib");
function getPublicKeyBuffer(publicKey, { compressed = true } = {}) {
    const res = noble_ecc_1.ecc.pointCompress(publicKey, compressed);
    if (res === null) {
        throw new Error('invalid public key');
    }
    const buffer = Buffer.from(res);
    assert.strictEqual(buffer.length, compressed ? 33 : 65);
    return buffer;
}
function toUncompressedPub(pubkey) {
    return getPublicKeyBuffer(pubkey, { compressed: false });
}
exports.toUncompressedPub = toUncompressedPub;
function toCompressedPub(pubkey) {
    return getPublicKeyBuffer(pubkey, { compressed: true });
}
exports.toCompressedPub = toCompressedPub;
/** create p2sh scripts with uncompressed pubkeys */
function createLegacySafeOutputScript2of3(pubkeys, network) {
    if (network) {
        if (!(0, networks_1.isBitcoin)(network)) {
            throw new Error(`unsupported network for legacy safe output script: ${network.coin}`);
        }
    }
    if (!(0, types_1.isTriple)(pubkeys)) {
        throw new Error(`must provide pubkey triple`);
    }
    pubkeys.forEach((key) => {
        if (key.length !== pubkeys[0].length) {
            throw new Error(`all pubkeys must have the same length`);
        }
        if (key.length !== 65 && key.length !== 33) {
            // V1 Safe BTC wallets could contain either uncompressed or compressed pubkeys
            throw new Error(`Unexpected key length ${key.length}, neither compressed nor uncompressed.`);
        }
    });
    const script2of3 = bitcoinjs.payments.p2ms({ m: 2, pubkeys });
    assert(script2of3.output);
    const scriptPubKey = bitcoinjs.payments.p2sh({ redeem: script2of3 });
    assert(scriptPubKey);
    assert(scriptPubKey.output);
    return {
        scriptPubKey: scriptPubKey.output,
        redeemScript: script2of3.output,
    };
}
exports.createLegacySafeOutputScript2of3 = createLegacySafeOutputScript2of3;
