// @bitgo/abstract-utxo/dist/src/abstractUtxoCoin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractUtxoCoin = exports.AbstractUtxoCoinWallet = void 0;
/**
 * @prettier
 */
const utxolib = require("@bitgo/utxo-lib");
const utxo_lib_1 = require("@bitgo/utxo-lib");
const assert = require("assert");
const bitcoinMessage = require("bitcoinjs-message");
const crypto_1 = require("crypto");
const debugLib = require("debug");
const _ = require("lodash");
const bignumber_js_1 = require("bignumber.js");
const recovery_1 = require("./recovery");
const sdk_core_1 = require("@bitgo/sdk-core");
const parseOutput_1 = require("./parseOutput");
const debug = debugLib('bitgo:v2:utxo');
const replayProtection_1 = require("./replayProtection");
const sign_1 = require("./sign");
const config_1 = require("./config");
const transaction_1 = require("./transaction");
const { getExternalChainCode, isChainCode, scriptTypeForChain, outputScripts } = utxo_lib_1.bitgo;
class AbstractUtxoCoinWallet extends sdk_core_1.Wallet {
    constructor(bitgo, baseCoin, walletData) {
        super(bitgo, baseCoin, walletData);
    }
}
exports.AbstractUtxoCoinWallet = AbstractUtxoCoinWallet;
class AbstractUtxoCoin extends sdk_core_1.BaseCoin {
    constructor(bitgo, network, amountType = 'number') {
        super(bitgo);
        if (!utxolib.isValidNetwork(network)) {
            throw new Error('invalid network: please make sure to use the same version of ' +
                '@bitgo/utxo-lib as this library when initializing an instance of this class');
        }
        this.amountType = amountType;
        this._network = network;
    }
    get network() {
        return this._network;
    }
    sweepWithSendMany() {
        return true;
    }
    /** @deprecated */
    static get validAddressTypes() {
        return [...outputScripts.scriptTypes2Of3];
    }
    /**
     * Returns the factor between the base unit and its smallest subdivison
     * @return {number}
     */
    getBaseFactor() {
        return 1e8;
    }
    /**
     * @deprecated
     */
    getCoinLibrary() {
        return utxolib;
    }
    /**
     * Check if an address is valid
     * @param address
     * @param param
     */
    isValidAddress(address, param) {
        if (typeof param === 'boolean' && param) {
            throw new Error('deprecated');
        }
        const formats = param && param.anyFormat ? undefined : ['default'];
        try {
            const script = utxolib.addressFormat.toOutputScriptTryFormats(address, this.network, formats);
            return address === utxolib.address.fromOutputScript(script, this.network);
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Return boolean indicating whether input is valid public key for the coin.
     *
     * @param {String} pub the pub to be checked
     * @returns {Boolean} is it valid?
     */
    isValidPub(pub) {
        try {
            return utxo_lib_1.bip32.fromBase58(pub).isNeutered();
        }
        catch (e) {
            return false;
        }
    }
    /**
     * Get the latest block height
     * @param reqId
     */
    async getLatestBlockHeight(reqId) {
        if (reqId) {
            this.bitgo.setRequestTracer(reqId);
        }
        const chainhead = await this.bitgo.get(this.url('/public/block/latest')).result();
        return chainhead.height;
    }
    /**
     * Run custom coin logic after a transaction prebuild has been received from BitGo
     * @param prebuild
     */
    async postProcessPrebuild(prebuild) {
        if (_.isUndefined(prebuild.txHex)) {
            throw new Error('missing required txPrebuild property txHex');
        }
        const tx = utxo_lib_1.bitgo.isPsbt(prebuild.txHex)
            ? utxo_lib_1.bitgo.createPsbtFromHex(prebuild.txHex, this.network)
            : this.createTransactionFromHex(prebuild.txHex);
        if (_.isUndefined(prebuild.blockHeight)) {
            prebuild.blockHeight = (await this.getLatestBlockHeight());
        }
        return _.extend({}, prebuild, { txHex: tx.toHex() });
    }
    /**
     * @param first
     * @param second
     * @returns {Array} All outputs that are in the first array but not in the second
     */
    static outputDifference(first, second) {
        const keyFunc = ({ address, amount }) => `${address}:${amount}`;
        const groupedOutputs = _.groupBy(first, keyFunc);
        second.forEach((output) => {
            const group = groupedOutputs[keyFunc(output)];
            if (group) {
                group.pop();
            }
        });
        return _.flatten(_.values(groupedOutputs));
    }
    /**
     * Determine an address' type based on its witness and redeem script presence
     * @param addressDetails
     */
    static inferAddressType(addressDetails) {
        return isChainCode(addressDetails.chain) ? scriptTypeForChain(addressDetails.chain) : null;
    }
    createTransactionFromHex(hex) {
        return utxolib.bitgo.createTransactionFromHex(hex, this.network, this.amountType);
    }
    /**
     * Extract and fill transaction details such as internal/change spend, external spend (explicit vs. implicit), etc.
     * @param params
     * @returns {*}
     */
    async parseTransaction(params) {
        const { txParams, txPrebuild, wallet, verification = {}, reqId } = params;
        if (!_.isUndefined(verification.disableNetworking) && !_.isBoolean(verification.disableNetworking)) {
            throw new Error('verification.disableNetworking must be a boolean');
        }
        const disableNetworking = verification.disableNetworking;
        const fetchKeychains = async (wallet) => {
            return (0, sdk_core_1.promiseProps)({
                user: this.keychains().get({ id: wallet.keyIds()[sdk_core_1.KeyIndices.USER], reqId }),
                backup: this.keychains().get({ id: wallet.keyIds()[sdk_core_1.KeyIndices.BACKUP], reqId }),
                bitgo: this.keychains().get({ id: wallet.keyIds()[sdk_core_1.KeyIndices.BITGO], reqId }),
            });
        };
        // obtain the keychains and key signatures
        let keychains = verification.keychains;
        if (!keychains) {
            if (disableNetworking) {
                throw new Error('cannot fetch keychains without networking');
            }
            keychains = await fetchKeychains(wallet);
        }
        if (!keychains || !keychains.user || !keychains.backup || !keychains.bitgo) {
            throw new Error('keychains are required, but could not be fetched');
        }
        const keychainArray = [keychains.user, keychains.backup, keychains.bitgo];
        const keySignatures = _.get(wallet, '_wallet.keySignatures', {});
        if (_.isUndefined(txPrebuild.txHex)) {
            throw new Error('missing required txPrebuild property txHex');
        }
        // obtain all outputs
        const explanation = await this.explainTransaction({
            txHex: txPrebuild.txHex,
            txInfo: txPrebuild.txInfo,
            pubs: keychainArray.map((k) => k.pub),
        });
        const allOutputs = [...explanation.outputs, ...explanation.changeOutputs];
        let expectedOutputs;
        if (txParams.rbfTxIds) {
            assert(txParams.rbfTxIds.length === 1);
            const txToBeReplaced = await wallet.getTransaction({ txHash: txParams.rbfTxIds[0], includeRbf: true });
            expectedOutputs = txToBeReplaced.outputs
                .filter((output) => output.wallet !== wallet.id()) // For self-sends, the walletId will be the same as the wallet's id
                .map((output) => {
                return { amount: BigInt(output.valueString), address: this.canonicalAddress(output.address) };
            });
        }
        else {
            // verify that each recipient from txParams has their own output
            expectedOutputs = _.get(txParams, 'recipients', []).map((output) => {
                return { ...output, address: this.canonicalAddress(output.address) };
            });
        }
        const missingOutputs = AbstractUtxoCoin.outputDifference(expectedOutputs, allOutputs);
        // get the keychains from the custom change wallet if needed
        let customChange;
        const { customChangeWalletId = undefined } = wallet.coinSpecific() || {};
        if (customChangeWalletId) {
            // fetch keychains from custom change wallet for deriving addresses.
            // These keychains should be signed and this should be verified in verifyTransaction
            const customChangeKeySignatures = wallet._wallet.customChangeKeySignatures;
            const customChangeWallet = await this.wallets().get({ id: customChangeWalletId });
            const customChangeKeys = await fetchKeychains(customChangeWallet);
            if (!customChangeKeys) {
                throw new Error('failed to fetch keychains for custom change wallet');
            }
            if (customChangeKeys.user && customChangeKeys.backup && customChangeKeys.bitgo && customChangeWallet) {
                const customChangeKeychains = [
                    customChangeKeys.user,
                    customChangeKeys.backup,
                    customChangeKeys.bitgo,
                ];
                customChange = {
                    keys: customChangeKeychains,
                    signatures: [
                        customChangeKeySignatures.user,
                        customChangeKeySignatures.backup,
                        customChangeKeySignatures.bitgo,
                    ],
                };
            }
        }
        /**
         * Loop through all the outputs and classify each of them as either internal spends
         * or external spends by setting the "external" property to true or false on the output object.
         */
        const allOutputDetails = await Promise.all(allOutputs.map((currentOutput) => {
            return (0, parseOutput_1.parseOutput)({
                currentOutput,
                coin: this,
                txPrebuild,
                verification,
                keychainArray,
                wallet,
                txParams,
                customChange,
                reqId,
            });
        }));
        const needsCustomChangeKeySignatureVerification = allOutputDetails.some((output) => output.needsCustomChangeKeySignatureVerification);
        const changeOutputs = _.filter(allOutputDetails, { external: false });
        // these are all the outputs that were not originally explicitly specified in recipients
        // ideally change outputs or a paygo output that might have been added
        const implicitOutputs = AbstractUtxoCoin.outputDifference(allOutputDetails, expectedOutputs);
        const explicitOutputs = AbstractUtxoCoin.outputDifference(allOutputDetails, implicitOutputs);
        // these are all the non-wallet outputs that had been originally explicitly specified in recipients
        const explicitExternalOutputs = _.filter(explicitOutputs, { external: true });
        // this is the sum of all the originally explicitly specified non-wallet output values
        const explicitExternalSpendAmount = utxolib.bitgo.toTNumber(explicitExternalOutputs.reduce((sum, o) => sum + BigInt(o.amount), BigInt(0)), this.amountType);
        /**
         * The calculation of the implicit external spend amount pertains to verifying the pay-as-you-go-fee BitGo
         * automatically applies to transactions sending money out of the wallet. The logic is fairly straightforward
         * in that we compare the external spend amount that was specified explicitly by the user to the portion
         * that was specified implicitly. To protect customers from people tampering with the transaction outputs, we
         * define a threshold for the maximum percentage of the implicit external spend in relation to the explicit
         * external spend.
         */
        // make sure that all the extra addresses are change addresses
        // get all the additional external outputs the server added and calculate their values
        const implicitExternalOutputs = _.filter(implicitOutputs, { external: true });
        const implicitExternalSpendAmount = utxolib.bitgo.toTNumber(implicitExternalOutputs.reduce((sum, o) => sum + BigInt(o.amount), BigInt(0)), this.amountType);
        return {
            keychains,
            keySignatures,
            outputs: allOutputDetails,
            missingOutputs,
            explicitExternalOutputs,
            implicitExternalOutputs,
            changeOutputs,
            explicitExternalSpendAmount,
            implicitExternalSpendAmount,
            needsCustomChangeKeySignatureVerification,
            customChange,
        };
    }
    /**
     * Decrypt the wallet's user private key and verify that the claimed public key matches
     * @param {VerifyUserPublicKeyOptions} params
     * @return {boolean}
     * @protected
     */
    verifyUserPublicKey(params) {
        const { userKeychain, txParams, disableNetworking } = params;
        if (!userKeychain) {
            throw new Error('user keychain is required');
        }
        const userPub = userKeychain.pub;
        // decrypt the user private key, so we can verify that the claimed public key is a match
        let userPrv = userKeychain.prv;
        if (!userPrv && txParams.walletPassphrase) {
            userPrv = (0, sdk_core_1.decryptKeychainPrivateKey)(this.bitgo, userKeychain, txParams.walletPassphrase);
        }
        if (!userPrv) {
            const errorMessage = 'user private key unavailable for verification';
            if (disableNetworking) {
                console.log(errorMessage);
                return false;
            }
            else {
                throw new Error(errorMessage);
            }
        }
        else {
            const userPrivateKey = utxo_lib_1.bip32.fromBase58(userPrv);
            if (userPrivateKey.toBase58() === userPrivateKey.neutered().toBase58()) {
                throw new Error('user private key is only public');
            }
            if (userPrivateKey.neutered().toBase58() !== userPub) {
                throw new Error('user private key does not match public key');
            }
        }
        return true;
    }
    /**
     * Verify signatures produced by the user key over the backup and bitgo keys.
     *
     * If set, these signatures ensure that the wallet keys cannot be changed after the wallet has been created.
     * @param {VerifyKeySignaturesOptions} params
     * @return {{backup: boolean, bitgo: boolean}}
     */
    verifyKeySignature(params) {
        // first, let's verify the integrity of the user key, whose public key is used for subsequent verifications
        const { userKeychain, keychainToVerify, keySignature } = params;
        if (!userKeychain) {
            throw new Error('user keychain is required');
        }
        if (!keychainToVerify) {
            throw new Error('keychain to verify is required');
        }
        if (!keySignature) {
            throw new Error('key signature is required');
        }
        // verify the signature against the user public key
        assert(userKeychain.pub);
        const publicKey = utxo_lib_1.bip32.fromBase58(userKeychain.pub).publicKey;
        // Due to interface of `bitcoinMessage`, we need to convert the public key to an address.
        // Note that this address has no relationship to on-chain transactions. We are
        // only interested in the address as a representation of the public key.
        const signingAddress = utxolib.address.toBase58Check(utxolib.crypto.hash160(publicKey), utxolib.networks.bitcoin.pubKeyHash, 
        // we do not pass `this.network` here because it would fail for zcash
        // the bitcoinMessage library decodes the address and throws away the first byte
        // because zcash has a two-byte prefix, verify() decodes zcash addresses to an invalid pubkey hash
        utxolib.networks.bitcoin);
        // BG-5703: use BTC mainnet prefix for all key signature operations
        // (this means do not pass a prefix parameter, and let it use the default prefix instead)
        assert(keychainToVerify.pub);
        try {
            return bitcoinMessage.verify(keychainToVerify.pub, signingAddress, Buffer.from(keySignature, 'hex'));
        }
        catch (e) {
            debug('error thrown from bitcoinmessage while verifying key signature', e);
            return false;
        }
    }
    /**
     * Verify signatures against the user private key over the change wallet extended keys
     * @param {ParsedTransaction} tx
     * @param {Keychain} userKeychain
     * @return {boolean}
     * @protected
     */
    verifyCustomChangeKeySignatures(tx, userKeychain) {
        if (!tx.customChange) {
            throw new Error('parsed transaction is missing required custom change verification data');
        }
        if (!Array.isArray(tx.customChange.keys) || !Array.isArray(tx.customChange.signatures)) {
            throw new Error('customChange property is missing keys or signatures');
        }
        for (const keyIndex of [sdk_core_1.KeyIndices.USER, sdk_core_1.KeyIndices.BACKUP, sdk_core_1.KeyIndices.BITGO]) {
            const keychainToVerify = tx.customChange.keys[keyIndex];
            const keySignature = tx.customChange.signatures[keyIndex];
            if (!keychainToVerify) {
                throw new Error(`missing required custom change ${sdk_core_1.KeyIndices[keyIndex].toLowerCase()} keychain public key`);
            }
            if (!keySignature) {
                throw new Error(`missing required custom change ${sdk_core_1.KeyIndices[keyIndex].toLowerCase()} keychain signature`);
            }
            if (!this.verifyKeySignature({
                userKeychain: userKeychain,
                keychainToVerify: keychainToVerify,
                keySignature,
            })) {
                debug('failed to verify custom change %s key signature!', sdk_core_1.KeyIndices[keyIndex].toLowerCase());
                return false;
            }
        }
        return true;
    }
    /**
     * Get the maximum percentage limit for pay-as-you-go outputs
     *
     * @protected
     */
    getPayGoLimit(allowPaygoOutput) {
        // allowing paygo outputs needs to be the default behavior, so only disallow paygo outputs if the
        // relevant verification option is both set and false
        if (!_.isNil(allowPaygoOutput) && !allowPaygoOutput) {
            return 0;
        }
        // 150 basis points is the absolute permitted maximum if paygo outputs are allowed
        return 0.015;
    }
    /**
     * Verify that a transaction prebuild complies with the original intention
     *
     * @param params
     * @param params.txParams params object passed to send
     * @param params.txPrebuild prebuild object returned by server
     * @param params.txPrebuild.txHex prebuilt transaction's txHex form
     * @param params.wallet Wallet object to obtain keys to verify against
     * @param params.verification Object specifying some verification parameters
     * @param params.verification.disableNetworking Disallow fetching any data from the internet for verification purposes
     * @param params.verification.keychains Pass keychains manually rather than fetching them by id
     * @param params.verification.addresses Address details to pass in for out-of-band verification
     * @returns {boolean}
     */
    async verifyTransaction(params) {
        var _a;
        const { txParams, txPrebuild, wallet, verification = { allowPaygoOutput: true }, reqId } = params;
        const isPsbt = txPrebuild.txHex && utxo_lib_1.bitgo.isPsbt(txPrebuild.txHex);
        if (isPsbt && ((_a = txPrebuild.txInfo) === null || _a === void 0 ? void 0 : _a.unspents)) {
            throw new Error('should not have unspents in txInfo for psbt');
        }
        const disableNetworking = !!verification.disableNetworking;
        const parsedTransaction = await this.parseTransaction({
            txParams,
            txPrebuild,
            wallet,
            verification,
            reqId,
        });
        const keychains = parsedTransaction.keychains;
        // verify that the claimed user public key corresponds to the wallet's user private key
        let userPublicKeyVerified = false;
        try {
            // verify the user public key matches the private key - this will throw if there is no match
            userPublicKeyVerified = this.verifyUserPublicKey({ userKeychain: keychains.user, disableNetworking, txParams });
        }
        catch (e) {
            debug('failed to verify user public key!', e);
        }
        // let's verify these keychains
        const keySignatures = parsedTransaction.keySignatures;
        if (!_.isEmpty(keySignatures)) {
            const verify = (key, pub) => {
                if (!keychains.user || !keychains.user.pub) {
                    throw new Error('missing user keychain');
                }
                return this.verifyKeySignature({
                    userKeychain: keychains.user,
                    keychainToVerify: key,
                    keySignature: pub,
                });
            };
            const isBackupKeySignatureValid = verify(keychains.backup, keySignatures.backupPub);
            const isBitgoKeySignatureValid = verify(keychains.bitgo, keySignatures.bitgoPub);
            if (!isBackupKeySignatureValid || !isBitgoKeySignatureValid) {
                throw new Error('secondary public key signatures invalid');
            }
            debug('successfully verified backup and bitgo key signatures');
        }
        else if (!disableNetworking) {
            // these keys were obtained online and their signatures were not verified
            // this could be dangerous
            console.log('unsigned keys obtained online are being used for address verification');
        }
        if (parsedTransaction.needsCustomChangeKeySignatureVerification) {
            if (!keychains.user || !userPublicKeyVerified) {
                throw new Error('transaction requires verification of user public key, but it was unable to be verified');
            }
            const customChangeKeySignaturesVerified = this.verifyCustomChangeKeySignatures(parsedTransaction, keychains.user);
            if (!customChangeKeySignaturesVerified) {
                throw new Error('transaction requires verification of custom change key signatures, but they were unable to be verified');
            }
            debug('successfully verified user public key and custom change key signatures');
        }
        const missingOutputs = parsedTransaction.missingOutputs;
        if (missingOutputs.length !== 0) {
            // there are some outputs in the recipients list that have not made it into the actual transaction
            throw new Error('expected outputs missing in transaction prebuild');
        }
        const intendedExternalSpend = parsedTransaction.explicitExternalSpendAmount;
        // this is a limit we impose for the total value that is amended to the transaction beyond what was originally intended
        const payAsYouGoLimit = new bignumber_js_1.default(this.getPayGoLimit(verification.allowPaygoOutput)).multipliedBy(intendedExternalSpend.toString());
        /*
        Some explanation for why we're doing what we're doing:
        Some customers will have an output to BitGo's PAYGo wallet added to their transaction, and we need to account for
        it here. To protect someone tampering with the output to make it send more than it should to BitGo, we define a
        threshold for the output's value above which we'll throw an error, because the paygo output should never be that
        high.
         */
        // make sure that all the extra addresses are change addresses
        // get all the additional external outputs the server added and calculate their values
        const nonChangeAmount = new bignumber_js_1.default(parsedTransaction.implicitExternalSpendAmount.toString());
        debug('Intended spend is %s, Non-change amount is %s, paygo limit is %s', intendedExternalSpend.toString(), nonChangeAmount.toString(), payAsYouGoLimit.toString());
        // the additional external outputs can only be BitGo's pay-as-you-go fee, but we cannot verify the wallet address
        if (nonChangeAmount.gt(payAsYouGoLimit)) {
            // there are some addresses that are outside the scope of intended recipients that are not change addresses
            throw new Error('prebuild attempts to spend to unintended external recipients');
        }
        const allOutputs = parsedTransaction.outputs;
        if (!txPrebuild.txHex) {
            throw new Error(`txPrebuild.txHex not set`);
        }
        const inputs = isPsbt
            ? (0, transaction_1.getPsbtTxInputs)(txPrebuild.txHex, this.network).map((v) => ({
                ...v,
                value: utxo_lib_1.bitgo.toTNumber(v.value, this.amountType),
            }))
            : await (0, transaction_1.getTxInputs)({ txPrebuild, bitgo: this.bitgo, coin: this, disableNetworking, reqId });
        // coins (doge) that can exceed number limits (and thus will use bigint) will have the `valueString` field
        const inputAmount = inputs.reduce((sum, i) => sum + BigInt(this.amountType === 'bigint' ? i.valueString : i.value), BigInt(0));
        const outputAmount = allOutputs.reduce((sum, o) => sum + BigInt(o.amount), BigInt(0));
        const fee = inputAmount - outputAmount;
        if (fee < 0) {
            throw new Error(`attempting to spend ${outputAmount} satoshis, which exceeds the input amount (${inputAmount} satoshis) by ${-fee}`);
        }
        return true;
    }
    /**
     * Make sure an address is valid and throw an error if it's not.
     * @param params.address The address string on the network
     * @param params.addressType
     * @param params.keychains Keychain objects with xpubs
     * @param params.coinSpecific Coin-specific details for the address such as a witness script
     * @param params.chain Derivation chain
     * @param params.index Derivation index
     * @throws {InvalidAddressError}
     * @throws {InvalidAddressDerivationPropertyError}
     * @throws {UnexpectedAddressError}
     */
    async isWalletAddress(params) {
        const { address, addressType, keychains, coinSpecific, chain, index } = params;
        if (!this.isValidAddress(address)) {
            throw new sdk_core_1.InvalidAddressError(`invalid address: ${address}`);
        }
        if ((_.isUndefined(chain) && _.isUndefined(index)) || !(_.isFinite(chain) && _.isFinite(index))) {
            throw new sdk_core_1.InvalidAddressDerivationPropertyError(`address validation failure: invalid chain (${chain}) or index (${index})`);
        }
        if (!_.isObject(coinSpecific)) {
            throw new sdk_core_1.InvalidAddressVerificationObjectPropertyError('address validation failure: coinSpecific field must be an object');
        }
        if (!keychains) {
            throw new Error('missing required param keychains');
        }
        const expectedAddress = this.generateAddress({
            addressType: addressType,
            keychains,
            threshold: 2,
            chain,
            index,
        });
        if (expectedAddress.address !== address) {
            throw new sdk_core_1.UnexpectedAddressError(`address validation failure: expected ${expectedAddress.address} but got ${address}`);
        }
        return true;
    }
    /**
     * Indicates whether coin supports a block target
     * @returns {boolean}
     */
    supportsBlockTarget() {
        return true;
    }
    /**
     * @param addressType
     * @returns true iff coin supports spending from unspentType
     */
    supportsAddressType(addressType) {
        return utxolib.bitgo.outputScripts.isSupportedScriptType(this.network, addressType);
    }
    /**
     * @param chain
     * @return true iff coin supports spending from chain
     */
    supportsAddressChain(chain) {
        return isChainCode(chain) && this.supportsAddressType(utxolib.bitgo.scriptTypeForChain(chain));
    }
    keyIdsForSigning() {
        return [sdk_core_1.KeyIndices.USER, sdk_core_1.KeyIndices.BACKUP, sdk_core_1.KeyIndices.BITGO];
    }
    /**
     * TODO(BG-11487): Remove addressType, segwit, and bech32 params in SDKv6
     * Generate an address for a wallet based on a set of configurations
     * @param params.addressType {string}   Deprecated
     * @param params.keychains   {[object]} Array of objects with xpubs
     * @param params.threshold   {number}   Minimum number of signatures
     * @param params.chain       {number}   Derivation chain (see https://github.com/BitGo/unspents/blob/master/src/codes.ts for
     *                                                 the corresponding address type of a given chain code)
     * @param params.index       {number}   Derivation index
     * @param params.segwit      {boolean}  Deprecated
     * @param params.bech32      {boolean}  Deprecated
     * @returns {{chain: number, index: number, coin: number, coinSpecific: {outputScript, redeemScript}}}
     */
    generateAddress(params) {
        const { keychains, threshold, chain, index, segwit = false, bech32 = false } = params;
        let derivationChain = getExternalChainCode('p2sh');
        if (_.isNumber(chain) && _.isInteger(chain) && isChainCode(chain)) {
            derivationChain = chain;
        }
        function convertFlagsToAddressType() {
            if (isChainCode(chain)) {
                return utxolib.bitgo.scriptTypeForChain(chain);
            }
            if (_.isBoolean(segwit) && segwit) {
                return 'p2shP2wsh';
            }
            else if (_.isBoolean(bech32) && bech32) {
                return 'p2wsh';
            }
            else {
                return 'p2sh';
            }
        }
        const addressType = params.addressType || convertFlagsToAddressType();
        if (addressType !== utxolib.bitgo.scriptTypeForChain(derivationChain)) {
            throw new sdk_core_1.AddressTypeChainMismatchError(addressType, derivationChain);
        }
        if (!this.supportsAddressType(addressType)) {
            switch (addressType) {
                case 'p2sh':
                    throw new Error(`internal error: p2sh should always be supported`);
                case 'p2shP2wsh':
                    throw new sdk_core_1.P2shP2wshUnsupportedError();
                case 'p2wsh':
                    throw new sdk_core_1.P2wshUnsupportedError();
                case 'p2tr':
                    throw new sdk_core_1.P2trUnsupportedError();
                case 'p2trMusig2':
                    throw new sdk_core_1.P2trMusig2UnsupportedError();
                default:
                    throw new sdk_core_1.UnsupportedAddressTypeError();
            }
        }
        let signatureThreshold = 2;
        if (_.isInteger(threshold)) {
            signatureThreshold = threshold;
            if (signatureThreshold <= 0) {
                throw new Error('threshold has to be positive');
            }
            if (signatureThreshold > keychains.length) {
                throw new Error('threshold cannot exceed number of keys');
            }
        }
        let derivationIndex = 0;
        if (_.isInteger(index) && index > 0) {
            derivationIndex = index;
        }
        const path = '0/0/' + derivationChain + '/' + derivationIndex;
        const hdNodes = keychains.map(({ pub }) => utxo_lib_1.bip32.fromBase58(pub));
        const derivedKeys = hdNodes.map((hdNode) => hdNode.derivePath((0, sdk_core_1.sanitizeLegacyPath)(path)).publicKey);
        const { outputScript, redeemScript, witnessScript, address } = this.createMultiSigAddress(addressType, signatureThreshold, derivedKeys);
        return {
            address,
            chain: derivationChain,
            index: derivationIndex,
            coin: this.getChain(),
            coinSpecific: {
                outputScript: outputScript.toString('hex'),
                redeemScript: redeemScript && redeemScript.toString('hex'),
                witnessScript: witnessScript && witnessScript.toString('hex'),
            },
            addressType,
        };
    }
    /**
     * @returns input psbt added with deterministic MuSig2 nonce for bitgo key for each MuSig2 inputs.
     * @param psbtHex all MuSig2 inputs should contain user MuSig2 nonce
     * @param walletId
     */
    async signPsbt(psbtHex, walletId) {
        const params = { psbt: psbtHex };
        return await this.bitgo
            .post(this.url('/wallet/' + walletId + '/tx/signpsbt'))
            .send(params)
            .result();
    }
    /**
     * @returns input psbt added with deterministic MuSig2 nonce for bitgo key for each MuSig2 inputs from OVC.
     * @param ovcJson JSON object provided by OVC with fields psbtHex and walletId
     */
    async signPsbtFromOVC(ovcJson) {
        assert(ovcJson['psbtHex'], 'ovcJson must contain psbtHex');
        assert(ovcJson['walletId'], 'ovcJson must contain walletId');
        const psbt = (await this.signPsbt(ovcJson['psbtHex'], ovcJson['walletId'])).psbt;
        assert(psbt, 'psbt not found');
        return _.extend(ovcJson, { txHex: psbt });
    }
    /**
     * Assemble keychain and half-sign prebuilt transaction
     * @param params - {@see SignTransactionOptions}
     * @returns {Promise<SignedTransaction | HalfSignedUtxoTransaction>}
     */
    async signTransaction(params) {
        var _a, _b, _c;
        const txPrebuild = params.txPrebuild;
        if (_.isUndefined(txPrebuild) || !_.isObject(txPrebuild)) {
            if (!_.isUndefined(txPrebuild) && !_.isObject(txPrebuild)) {
                throw new Error(`txPrebuild must be an object, got type ${typeof txPrebuild}`);
            }
            throw new Error('missing txPrebuild parameter');
        }
        let tx = utxo_lib_1.bitgo.isPsbt(txPrebuild.txHex)
            ? utxo_lib_1.bitgo.createPsbtFromHex(txPrebuild.txHex, this.network)
            : this.createTransactionFromHex(txPrebuild.txHex);
        const isTxWithKeyPathSpendInput = tx instanceof utxo_lib_1.bitgo.UtxoPsbt && utxo_lib_1.bitgo.isTransactionWithKeyPathSpendInput(tx);
        let isLastSignature = false;
        if (_.isBoolean(params.isLastSignature)) {
            // We can only be the first signature on a transaction with taproot key path spend inputs because
            // we require the secret nonce in the cache of the first signer, which is impossible to retrieve if
            // deserialized from a hex.
            if (params.isLastSignature && isTxWithKeyPathSpendInput) {
                throw new Error('Cannot be last signature on a transaction with key path spend inputs');
            }
            // if build is called instead of buildIncomplete, no signature placeholders are left in the sig script
            isLastSignature = params.isLastSignature;
        }
        const getSignerKeychain = () => {
            const userPrv = params.prv;
            if (_.isUndefined(userPrv) || !_.isString(userPrv)) {
                if (!_.isUndefined(userPrv)) {
                    throw new Error(`prv must be a string, got type ${typeof userPrv}`);
                }
                throw new Error('missing prv parameter to sign transaction');
            }
            const signerKeychain = utxo_lib_1.bip32.fromBase58(userPrv, utxolib.networks.bitcoin);
            if (signerKeychain.isNeutered()) {
                throw new Error('expected user private key but received public key');
            }
            debug(`Here is the public key of the xprv you used to sign: ${signerKeychain.neutered().toBase58()}`);
            return signerKeychain;
        };
        const setSignerMusigNonceWithOverride = (psbt, signerKeychain, nonSegwitOverride) => {
            utxolib.bitgo.withUnsafeNonSegwit(psbt, () => psbt.setAllInputsMusig2NonceHD(signerKeychain), nonSegwitOverride);
        };
        let signerKeychain;
        if (tx instanceof utxo_lib_1.bitgo.UtxoPsbt && isTxWithKeyPathSpendInput) {
            switch (params.signingStep) {
                case 'signerNonce':
                    signerKeychain = getSignerKeychain();
                    setSignerMusigNonceWithOverride(tx, signerKeychain, !!params.allowNonSegwitSigningWithoutPrevTx);
                    AbstractUtxoCoin.PSBT_CACHE.set(tx.getUnsignedTx().getId(), tx);
                    return { txHex: tx.toHex() };
                case 'cosignerNonce':
                    assert(txPrebuild.walletId, 'walletId is required for MuSig2 bitgo nonce');
                    return { txHex: (await this.signPsbt(tx.toHex(), txPrebuild.walletId)).psbt };
                case 'signerSignature':
                    const txId = tx.getUnsignedTx().getId();
                    const psbt = AbstractUtxoCoin.PSBT_CACHE.get(txId);
                    assert(psbt, `Psbt is missing from txCache (cache size ${AbstractUtxoCoin.PSBT_CACHE.size}).
            This may be due to the request being routed to a different BitGo-Express instance that for signing step 'signerNonce'.`);
                    AbstractUtxoCoin.PSBT_CACHE.delete(txId);
                    tx = psbt.combine(tx);
                    break;
                default:
                    // this instance is not an external signer
                    assert(txPrebuild.walletId, 'walletId is required for MuSig2 bitgo nonce');
                    signerKeychain = getSignerKeychain();
                    setSignerMusigNonceWithOverride(tx, signerKeychain, !!params.allowNonSegwitSigningWithoutPrevTx);
                    const response = await this.signPsbt(tx.toHex(), txPrebuild.walletId);
                    tx.combine(utxo_lib_1.bitgo.createPsbtFromHex(response.psbt, this.network));
                    break;
            }
        }
        else {
            switch (params.signingStep) {
                case 'signerNonce':
                case 'cosignerNonce':
                    /**
                     * In certain cases, the caller of this method may not know whether the txHex contains a psbt with taproot key path spend input(s).
                     * Instead of throwing error, no-op and return the txHex. So that the caller can call this method in the same sequence.
                     */
                    return { txHex: tx.toHex() };
            }
        }
        if (signerKeychain === undefined) {
            signerKeychain = getSignerKeychain();
        }
        let signedTransaction;
        if (tx instanceof utxo_lib_1.bitgo.UtxoPsbt) {
            signedTransaction = (0, sign_1.signAndVerifyPsbt)(tx, signerKeychain, {
                isLastSignature,
                allowNonSegwitSigningWithoutPrevTx: params.allowNonSegwitSigningWithoutPrevTx,
            });
        }
        else {
            if (tx.ins.length !== ((_b = (_a = txPrebuild.txInfo) === null || _a === void 0 ? void 0 : _a.unspents) === null || _b === void 0 ? void 0 : _b.length)) {
                throw new Error('length of unspents array should equal to the number of transaction inputs');
            }
            if (!params.pubs || !(0, sdk_core_1.isTriple)(params.pubs)) {
                throw new Error(`must provide xpub array`);
            }
            const keychains = params.pubs.map((pub) => utxo_lib_1.bip32.fromBase58(pub));
            const cosignerPub = (_c = params.cosignerPub) !== null && _c !== void 0 ? _c : params.pubs[2];
            const cosignerKeychain = utxo_lib_1.bip32.fromBase58(cosignerPub);
            const walletSigner = new utxo_lib_1.bitgo.WalletUnspentSigner(keychains, signerKeychain, cosignerKeychain);
            signedTransaction = (0, sign_1.signAndVerifyWalletTransaction)(tx, txPrebuild.txInfo.unspents, walletSigner, {
                isLastSignature,
            });
        }
        return {
            txHex: signedTransaction.toBuffer().toString('hex'),
        };
    }
    /**
     * Sign a transaction with a custom signing function. Example use case is express external signer
     * @param customSigningFunction custom signing function that returns a single signed transaction
     * @param signTransactionParams parameters for custom signing function. Includes txPrebuild and pubs (for legacy tx only).
     *
     * @returns signed transaction as hex string
     */
    async signWithCustomSigningFunction(customSigningFunction, signTransactionParams) {
        const txHex = signTransactionParams.txPrebuild.txHex;
        assert(txHex, 'missing txHex parameter');
        const tx = utxo_lib_1.bitgo.isPsbt(txHex)
            ? utxo_lib_1.bitgo.createPsbtFromHex(txHex, this.network)
            : this.createTransactionFromHex(txHex);
        const isTxWithKeyPathSpendInput = tx instanceof utxo_lib_1.bitgo.UtxoPsbt && utxo_lib_1.bitgo.isTransactionWithKeyPathSpendInput(tx);
        if (!isTxWithKeyPathSpendInput) {
            return await customSigningFunction({ ...signTransactionParams, coin: this });
        }
        const getTxHex = (v) => {
            if ('txHex' in v) {
                return v.txHex;
            }
            throw new Error('txHex not found in signTransaction result');
        };
        const signerNonceTx = await customSigningFunction({
            ...signTransactionParams,
            signingStep: 'signerNonce',
            coin: this,
        });
        const { pubs } = signTransactionParams;
        assert(pubs === undefined || (0, sdk_core_1.isTriple)(pubs));
        const cosignerNonceTx = await this.signTransaction({
            ...signTransactionParams,
            pubs,
            txPrebuild: { ...signTransactionParams.txPrebuild, txHex: getTxHex(signerNonceTx) },
            signingStep: 'cosignerNonce',
        });
        return await customSigningFunction({
            ...signTransactionParams,
            txPrebuild: { ...signTransactionParams.txPrebuild, txHex: getTxHex(cosignerNonceTx) },
            signingStep: 'signerSignature',
            coin: this,
        });
    }
    /**
     * @param unspent
     * @returns {boolean}
     */
    isBitGoTaintedUnspent(unspent) {
        return (0, replayProtection_1.isReplayProtectionUnspent)(unspent, this.network);
    }
    /**
     * @deprecated - use utxolib.bitgo.getDefaultSigHash(network) instead
     * @returns {number}
     */
    get defaultSigHashType() {
        return utxolib.bitgo.getDefaultSigHash(this.network);
    }
    /**
     * @deprecated - use utxolib.bitcoin.verifySignature() instead
     */
    verifySignature(transaction, inputIndex, amount, verificationSettings = {}) {
        if (transaction.network !== this.network) {
            throw new Error(`network mismatch`);
        }
        return utxolib.bitgo.verifySignature(transaction, inputIndex, amount, {
            signatureIndex: verificationSettings.signatureIndex,
            publicKey: verificationSettings.publicKey ? Buffer.from(verificationSettings.publicKey, 'hex') : undefined,
        });
    }
    /**
     * Decompose a raw psbt/transaction into useful information, such as the total amounts,
     * change amounts, and transaction outputs.
     * @param params
     */
    async explainTransaction(params) {
        const { txHex } = params;
        if (typeof txHex !== 'string' || !txHex.match(/^([a-f0-9]{2})+$/i)) {
            throw new Error('invalid transaction hex, must be a valid hex string');
        }
        return utxolib.bitgo.isPsbt(txHex) ? (0, transaction_1.explainPsbt)(params, this.network) : (0, transaction_1.explainTx)(params, this);
    }
    /**
     * Create a multisig address of a given type from a list of keychains and a signing threshold
     * @param addressType
     * @param signatureThreshold
     * @param keys
     */
    createMultiSigAddress(addressType, signatureThreshold, keys) {
        const { scriptPubKey: outputScript, redeemScript, witnessScript, } = utxolib.bitgo.outputScripts.createOutputScript2of3(keys, addressType);
        return {
            outputScript,
            redeemScript,
            witnessScript,
            address: utxolib.address.fromOutputScript(outputScript, this.network),
        };
    }
    /**
     * @deprecated - use {@see backupKeyRecovery}
     * Builds a funds recovery transaction without BitGo
     * @param params - {@see backupKeyRecovery}
     */
    async recover(params) {
        return (0, recovery_1.backupKeyRecovery)(this, this.bitgo, params);
    }
    async recoverV1(params) {
        return (0, recovery_1.v1BackupKeyRecovery)(this, this.bitgo, params);
    }
    /**
     * Recover coin that was sent to wrong chain
     * @param params
     * @param params.txid The txid of the faulty transaction
     * @param params.recoveryAddress address to send recovered funds to
     * @param params.wallet the wallet that received the funds
     * @param params.recoveryCoin the coin type of the wallet that received the funds
     * @param params.signed return a half-signed transaction (default=true)
     * @param params.walletPassphrase the wallet passphrase
     * @param params.xprv the unencrypted xprv (used instead of wallet passphrase)
     * @param params.apiKey for utxo coins other than [BTC,TBTC] this is a Block Chair api key
     * @returns {*}
     */
    async recoverFromWrongChain(params) {
        const { txid, recoveryAddress, wallet, walletPassphrase, xprv, apiKey } = params;
        // params.recoveryCoin used to be params.coin, backwards compatibility
        const recoveryCoin = params.coin || params.recoveryCoin;
        if (!recoveryCoin) {
            throw new Error('missing required object recoveryCoin');
        }
        // signed should default to true, and only be disabled if explicitly set to false (not undefined)
        const signed = params.signed !== false;
        const sourceCoinFamily = this.getFamily();
        const recoveryCoinFamily = recoveryCoin.getFamily();
        const supportedRecoveryCoins = config_1.supportedCrossChainRecoveries[sourceCoinFamily];
        if (_.isUndefined(supportedRecoveryCoins) || !supportedRecoveryCoins.includes(recoveryCoinFamily)) {
            throw new Error(`Recovery of ${sourceCoinFamily} balances from ${recoveryCoinFamily} wallets is not supported.`);
        }
        return await (0, recovery_1.recoverCrossChain)(this.bitgo, {
            sourceCoin: this,
            recoveryCoin,
            walletId: wallet,
            txid,
            recoveryAddress,
            walletPassphrase: signed ? walletPassphrase : undefined,
            xprv: signed ? xprv : undefined,
            apiKey,
        });
    }
    /**
     * Generate bip32 key pair
     *
     * @param seed
     * @returns {Object} object with generated pub and prv
     */
    generateKeyPair(seed) {
        if (!seed) {
            // An extended private key has both a normal 256 bit private key and a 256
            // bit chain code, both of which must be random. 512 bits is therefore the
            // maximum entropy and gives us maximum security against cracking.
            seed = (0, crypto_1.randomBytes)(512 / 8);
        }
        const extendedKey = utxo_lib_1.bip32.fromSeed(seed);
        return {
            pub: extendedKey.neutered().toBase58(),
            prv: extendedKey.toBase58(),
        };
    }
    async getExtraPrebuildParams(buildParams) {
        let txFormat = buildParams.txFormat;
        let changeAddressType = buildParams.changeAddressType;
        const walletFlagMusigKp = buildParams.wallet.flag('musigKp') === 'true';
        // if the txFormat is not specified, we need to default to psbt for distributed custody wallets or testnet hot wallets
        if (buildParams.txFormat === undefined &&
            (buildParams.wallet.subType() === 'distributedCustody' ||
                ((0, utxo_lib_1.isTestnet)(this.network) && buildParams.wallet.type() === 'hot') ||
                // FIXME(BTC-776): default to psbt for all mainnet wallets in the future
                walletFlagMusigKp)) {
            txFormat = 'psbt';
        }
        // if the addressType is not specified, we need to default to p2trMusig2 for testnet hot wallets for staged rollout of p2trMusig2
        if (buildParams.addressType === undefined && // addressType is deprecated and replaced by `changeAddress`
            buildParams.changeAddressType === undefined &&
            buildParams.changeAddress === undefined &&
            buildParams.wallet.type() === 'hot') {
            changeAddressType = ['p2trMusig2', 'p2wsh', 'p2shP2wsh', 'p2sh', 'p2tr'];
        }
        return {
            txFormat,
            changeAddressType,
        };
    }
    preCreateBitGo(params) {
        return;
    }
    async presignTransaction(params) {
        return params;
    }
    async supplementGenerateWallet(walletParams, keychains) {
        return walletParams;
    }
    transactionDataAllowed() {
        return false;
    }
    valuelessTransferAllowed() {
        return false;
    }
    getRecoveryProvider(apiToken) {
        return (0, recovery_1.forCoin)(this.getChain(), apiToken);
    }
}
exports.AbstractUtxoCoin = AbstractUtxoCoin;
/**
 * Key Value: Unsigned tx id => PSBT
 * It is used to cache PSBTs with taproot key path (MuSig2) inputs during external express signer is activated.
 * Reason: MuSig2 signer secure nonce is cached in the UtxoPsbt object. It will be required during the signing step.
 * For more info, check SignTransactionOptions.signingStep
 *
 * TODO BTC-276: This cache may need to be done with LRU like memory safe caching if memory issues comes up.
 */
AbstractUtxoCoin.PSBT_CACHE = new Map();
