// @bitgo/sdk-coin-hash/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HashUtils = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const statics_1 = require("@bitgo/statics");
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const constants = __importStar(require("./constants"));
class HashUtils extends abstract_cosmos_1.CosmosUtils {
    constructor(networkType = statics_1.NetworkType.MAINNET) {
        super();
        this.networkType = networkType;
    }
    /** @inheritdoc */
    isValidAddress(address) {
        if (this.networkType === statics_1.NetworkType.TESTNET) {
            return this.isValidCosmosLikeAddressWithMemoId(address, constants.testnetAccountAddressRegex);
        }
        return this.isValidCosmosLikeAddressWithMemoId(address, constants.mainnetAccountAddressRegex);
    }
    /** @inheritdoc */
    isValidValidatorAddress(address) {
        if (this.networkType === statics_1.NetworkType.TESTNET) {
            return this.isValidBech32AddressMatchingRegex(address, constants.testnetValidatorAddressRegex);
        }
        return this.isValidBech32AddressMatchingRegex(address, constants.mainnetValidatorAddressRegex);
    }
    /** @inheritdoc */
    isValidContractAddress(address) {
        if (this.networkType === statics_1.NetworkType.TESTNET) {
            return this.isValidBech32AddressMatchingRegex(address, constants.testnetContractAddressRegex);
        }
        return this.isValidBech32AddressMatchingRegex(address, constants.mainnetContractAddressRegex);
    }
    /** @inheritdoc */
    validateAmount(amount) {
        const amountBig = (0, bignumber_js_1.default)(amount.amount);
        if (amountBig.isLessThanOrEqualTo(0)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid amount: ' + amount.amount);
        }
        if (!constants.validDenoms.find((denom) => denom === amount.denom)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid denom: ' + amount.denom);
        }
    }
}
exports.HashUtils = HashUtils;
const hashUtils = new HashUtils();
exports.default = hashUtils;
