// @bitgo/sdk-coin-stx/dist/src/lib/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const bn_js_1 = __importDefault(require("bn.js"));
const transactions_1 = require("@stacks/transactions");
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const utils_1 = require("./utils");
class TransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    initBuilder(tx) {
        const txData = tx.toJson();
        if (txData.payload === undefined) {
            throw new sdk_core_1.InvalidTransactionError('payload must not be undefined');
        }
        if (txData.payload.payloadType === transactions_1.PayloadType.TokenTransfer) {
            this.to(txData.payload.to);
            this.amount(txData.payload.amount);
            if (txData.payload.memo) {
                this.memo(txData.payload.memo);
            }
            super.initBuilder(tx);
        }
        else {
            throw new sdk_core_1.BuildTransactionError('Transaction should be transfer');
        }
    }
    /** @inheritdoc */
    async buildImplementation() {
        this._options = this.buildTokenTransferOptions();
        this.transaction.setTransactionType(sdk_core_1.TransactionType.Send);
        this.transaction.stxTransaction = await (0, transactions_1.makeUnsignedSTXTokenTransfer)(this._options);
        return await super.buildImplementation();
    }
    buildTokenTransferOptions() {
        const defaultOpts = {
            recipient: this._toAddress,
            amount: this._amount,
            memo: this._memo,
            anchorMode: this._anchorMode,
            network: this._network,
            fee: new bn_js_1.default(this._fee.fee),
            nonce: new bn_js_1.default(this._nonce),
        };
        if (this._fromPubKeys.length > 0) {
            if (this._fromPubKeys.length === 1) {
                return {
                    ...defaultOpts,
                    publicKey: this._fromPubKeys[0],
                };
            }
            else {
                return {
                    ...defaultOpts,
                    publicKeys: this._fromPubKeys,
                    numSignatures: this._numberSignatures,
                };
            }
        }
        else {
            throw new sdk_core_1.InvalidParameterValueError('supply at least 1 public key');
        }
    }
    // region Transfer fields
    /**
     * Set the destination address where the funds will be sent,
     * it may take the format `'<shard>.<realm>.<account>'` or `'<account>'`
     *
     * @param {string} address the address to transfer funds to
     * @returns {TransferBuilder} the builder with the new parameter set
     */
    to(address) {
        if (!(0, utils_1.isValidAddress)(address)) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid address');
        }
        this._toAddress = address;
        return this;
    }
    /**
     * Set the amount to be transferred
     *
     * @param {string} amount amount to transfer in tinyBars (there are 100,000,000 tinyBars in one Hbar)
     * @returns {TransferBuilder} the builder with the new parameter set
     */
    amount(amount) {
        if (!(0, utils_1.isValidAmount)(amount)) {
            throw new sdk_core_1.InvalidParameterValueError('Invalid amount');
        }
        this._amount = new bn_js_1.default(amount);
        return this;
    }
}
exports.TransferBuilder = TransferBuilder;
