// @bitgo/sdk-coin-stx/dist/src/lib/transaction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const transactions_1 = require("@stacks/transactions");
const sdk_core_1 = require("@bitgo/sdk-core");
const utils_1 = require("./utils");
const bn_js_1 = __importDefault(require("bn.js"));
class Transaction extends sdk_core_1.BaseTransaction {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    /** @inheritdoc */
    canSign(key) {
        return true;
    }
    async sign(keyPair, sigHash) {
        var _a;
        const keyPairs = keyPair instanceof Array ? keyPair : [keyPair];
        const signer = new transactions_1.TransactionSigner(this._stxTransaction);
        signer.checkOversign = false;
        signer.sigHash = (_a = sigHash !== null && sigHash !== void 0 ? sigHash : this._sigHash) !== null && _a !== void 0 ? _a : this._stxTransaction.verifyBegin();
        for (const kp of keyPairs) {
            const keys = kp.getKeys(kp.getCompressed());
            if (!keys.prv) {
                throw new sdk_core_1.SigningError('Missing private key');
            }
            const privKey = (0, transactions_1.createStacksPrivateKey)(keys.prv);
            signer.signOrigin(privKey);
            this._sigHash = signer.sigHash;
        }
    }
    async appendOrigin(pubKeyString) {
        const pubKeyStrings = pubKeyString instanceof Array ? pubKeyString : [pubKeyString];
        const signer = new transactions_1.TransactionSigner(this._stxTransaction);
        pubKeyStrings.forEach((pubKey) => {
            signer.appendOrigin((0, transactions_1.createStacksPublicKey)(pubKey));
        });
    }
    async signWithSignatures(signature, isMultiSig) {
        if (!signature) {
            throw new sdk_core_1.SigningError('Missing signatures');
        }
        const signatures = signature instanceof Array ? signature : [signature];
        if (!isMultiSig) {
            this._stxTransaction = this._stxTransaction.createTxWithSignature(signatures[0].data);
        }
        else {
            const authFields = signatures.map((sig) => (0, transactions_1.createTransactionAuthField)(transactions_1.PubKeyEncoding.Compressed, sig));
            this._stxTransaction.auth.spendingCondition.fields = this._stxTransaction.auth.spendingCondition.fields.concat(authFields);
        }
        if (signatures.length > 0) {
            this._sigHash = signatures[signatures.length - 1].sigHash;
        }
    }
    get signature() {
        if (this._stxTransaction && this._stxTransaction.auth.spendingCondition) {
            if ((0, transactions_1.isSingleSig)(this._stxTransaction.auth.spendingCondition)) {
                return [this._stxTransaction.auth.spendingCondition.signature.data];
            }
            else {
                const signatures = [];
                this._stxTransaction.auth.spendingCondition.fields.forEach((field) => {
                    if (field.contents.type === transactions_1.StacksMessageType.MessageSignature) {
                        signatures.push(field.contents.data);
                    }
                });
                return signatures;
            }
        }
        return [];
    }
    /** @inheritdoc */
    toJson() {
        if (!this._stxTransaction) {
            throw new sdk_core_1.ParseTransactionError('Empty transaction');
        }
        const result = {
            id: this._stxTransaction.txid(),
            fee: this._stxTransaction.auth.getFee().toString(10),
            from: (0, utils_1.getTxSenderAddress)(this._stxTransaction),
            nonce: this.getNonce(),
            payload: this.getPayloadData(),
        };
        return result;
    }
    getPayloadData() {
        if (this._stxTransaction.payload.payloadType === transactions_1.PayloadType.TokenTransfer) {
            const payload = this._stxTransaction.payload;
            const txPayload = {
                payloadType: transactions_1.PayloadType.TokenTransfer,
                // result.payload.memo will be padded with \u0000 up to
                // MEMO_MAX_LENGTH_BYTES as defined in @stacks/transactions
                memo: (0, utils_1.unpadMemo)(payload.memo.content),
                to: (0, transactions_1.addressToString)({
                    type: transactions_1.StacksMessageType.Address,
                    version: payload.recipient.address.version,
                    hash160: payload.recipient.address.hash160.toString(),
                }),
                amount: payload.amount.toString(),
            };
            return txPayload;
        }
        else if (this._stxTransaction.payload.payloadType === transactions_1.PayloadType.ContractCall) {
            const payload = this._stxTransaction.payload;
            const contractPayload = {
                payloadType: transactions_1.PayloadType.ContractCall,
                contractAddress: (0, transactions_1.addressToString)(payload.contractAddress),
                contractName: payload.contractName.content,
                functionName: payload.functionName.content,
                functionArgs: payload.functionArgs.map(utils_1.stringifyCv),
            };
            return contractPayload;
        }
        else {
            throw new sdk_core_1.NotSupported('payload type not supported');
        }
    }
    /**
     * Return the length of a transaction.  This is needed to calculate
     * the transaction fee.
     *
     * @returns {number} size in bytes of the serialized transaction
     */
    transactionSize() {
        return this._stxTransaction.serialize().length;
    }
    toBroadcastFormat() {
        if (!this._stxTransaction) {
            throw new sdk_core_1.ParseTransactionError('Empty transaction');
        }
        return this._stxTransaction.serialize().toString('hex');
    }
    get stxTransaction() {
        return this._stxTransaction;
    }
    set stxTransaction(t) {
        this._stxTransaction = t;
    }
    getNonce() {
        if (this._stxTransaction.auth.spendingCondition) {
            return Number(this._stxTransaction.auth.spendingCondition.nonce);
        }
        else {
            throw new sdk_core_1.InvalidTransactionError('spending condition is null');
        }
    }
    /**
     * Sets this transaction payload
     *
     * @param rawTransaction
     */
    fromRawTransaction(rawTransaction) {
        const raw = (0, utils_1.removeHexPrefix)(rawTransaction);
        try {
            this._stxTransaction = (0, transactions_1.deserializeTransaction)(transactions_1.BufferReader.fromBuffer(Buffer.from(raw, 'hex')));
        }
        catch (e) {
            throw new sdk_core_1.ParseTransactionError('Error parsing the raw transaction');
        }
        this.loadInputsAndOutputs();
    }
    /**
     * Set the transaction type
     *
     * @param {TransactionType} transactionType The transaction type to be set
     */
    setTransactionType(transactionType) {
        this._type = transactionType;
    }
    /**
     * Load the input and output data on this transaction using the transaction json
     * if there are outputs.
     */
    loadInputsAndOutputs() {
        const txJson = this.toJson();
        if (txJson.payload.payloadType === transactions_1.PayloadType.TokenTransfer) {
            if (txJson.payload.to && txJson.payload.amount) {
                this._outputs = [
                    {
                        address: txJson.payload.to,
                        value: txJson.payload.amount,
                        coin: this._coinConfig.name,
                    },
                ];
                this._inputs = [
                    {
                        address: txJson.from,
                        value: txJson.payload.amount,
                        coin: this._coinConfig.name,
                    },
                ];
            }
        }
        else if (txJson.payload.payloadType === transactions_1.PayloadType.ContractCall) {
            if (txJson.payload.contractAddress === this._coinConfig.network.sendmanymemoContractAddress) {
                const sendParams = (0, utils_1.functionArgsToSendParams)(this.stxTransaction.payload.functionArgs);
                const coin = this._coinConfig.name;
                const sum = sendParams.reduce((current, next) => current.add(new bn_js_1.default(next.amount)), new bn_js_1.default(0));
                this._outputs = sendParams.map((sendParam) => ({ address: sendParam.address, value: sendParam.amount, coin }));
                this._inputs = [{ address: txJson.from, value: sum.toString(), coin }];
            }
            else {
                this._outputs = [
                    {
                        address: txJson.payload.contractAddress,
                        value: '0',
                        coin: this._coinConfig.name,
                    },
                ];
                this._inputs = [
                    {
                        address: txJson.from,
                        value: '0',
                        coin: this._coinConfig.name,
                    },
                ];
            }
        }
    }
}
exports.Transaction = Transaction;
