// @bitgo/sdk-coin-algo/dist/src/lib/seedEncoding.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SeedEncoding = void 0;
const base32 = __importStar(require("hi-base32"));
const js_sha512_1 = __importDefault(require("js-sha512"));
const SEED_BYTES_LENGTH = 32;
const ALGORAND_SEED_BYTE_LENGTH = 36;
const ALGORAND_CHECKSUM_BYTE_LENGTH = 4;
const ALGORAND_SEED_LENGTH = 58;
class SeedEncoding {
    static genericHash(arr) {
        return js_sha512_1.default.sha512_256.array(arr);
    }
    /**
     * Checks if a seed is valid
     * @param {String} seed - encoded Algorand seed
     * @returns {Boolean} true if valid, false otherwise
     */
    static isValidSeed(seed) {
        if (seed.length !== ALGORAND_SEED_LENGTH) {
            return false;
        }
        // Try to decode
        let decoded;
        try {
            decoded = SeedEncoding.decode(seed);
        }
        catch (e) {
            return false;
        }
        // Compute checksum
        const checksum = SeedEncoding.genericHash(decoded.seed).slice(SEED_BYTES_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH, SEED_BYTES_LENGTH);
        // Check if the checksum and the seed are equal
        if (checksum.length !== decoded.checksum.length) {
            return false;
        }
        return checksum.every((val, i) => val === decoded.checksum[i]);
    }
    /**
     * Decode a seed
     *
     * @param seed
     * @return {{checksum: Uint8Array, seed: Uint8Array}}
     */
    static decode(seed) {
        // try to decode
        const decoded = base32.decode.asBytes(seed);
        // Sanity check
        if (decoded.length !== ALGORAND_SEED_BYTE_LENGTH)
            throw new Error('seed seems to be malformed');
        return {
            seed: new Uint8Array(decoded.slice(0, ALGORAND_SEED_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH)),
            checksum: new Uint8Array(decoded.slice(SEED_BYTES_LENGTH, ALGORAND_SEED_BYTE_LENGTH)),
        };
    }
    /**
     * Encode a secret key into a seed
     *
     * @param secretKey
     * @return {String} encoded seed
     */
    static encode(secretKey) {
        // get seed
        const seed = secretKey.slice(0, SEED_BYTES_LENGTH);
        // compute checksum
        const checksum = SeedEncoding.genericHash(seed).slice(SEED_BYTES_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH, SEED_BYTES_LENGTH);
        const arraySeed = new Uint8Array(seed.length + checksum.length);
        arraySeed.set(seed);
        arraySeed.set(checksum, seed.length);
        const encodedSeed = base32.encode(arraySeed);
        return encodedSeed.toString().slice(0, ALGORAND_SEED_LENGTH); // removing the extra '===='
    }
}
exports.SeedEncoding = SeedEncoding;
