// @bitgo/sdk-coin-algo/dist/src/lib/transferBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const algosdk_1 = __importDefault(require("algosdk"));
const transactionBuilder_1 = require("./transactionBuilder");
const txnSchema_1 = require("./txnSchema");
const utils_1 = __importDefault(require("./utils"));
const sdk_core_1 = require("@bitgo/sdk-core");
class TransferBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(coinConfig) {
        super(coinConfig);
    }
    buildAlgoTxn() {
        return algosdk_1.default.makePaymentTxnWithSuggestedParams(this._sender, this._to, this._amount, this._closeRemainderTo, this._note, this.suggestedParams, this._reKeyTo);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.Send;
    }
    fromImplementation(rawTransaction) {
        const tx = super.fromImplementation(rawTransaction);
        const algoTx = tx.getAlgoTransaction();
        if (algoTx) {
            this.amount(algoTx.amount || 0);
            this.to({ address: algosdk_1.default.encodeAddress(algoTx.to.publicKey) });
            if (algoTx.closeRemainderTo) {
                this.closeRemainderTo({ address: algosdk_1.default.encodeAddress(algoTx.closeRemainderTo.publicKey) });
            }
        }
        return tx;
    }
    validateRawTransaction(rawTransaction) {
        const { txn: algoTxn } = utils_1.default.decodeAlgoTxn(rawTransaction);
        if (algoTxn.type !== algosdk_1.default.TransactionType.pay) {
            throw new sdk_core_1.InvalidTransactionError(`Invalid Transaction Type: ${algoTxn.type}. Expected ${algosdk_1.default.TransactionType.pay}`);
        }
    }
    validateTransaction(transaction) {
        super.validateTransaction(transaction);
        const validationResult = txnSchema_1.TransferTransactionSchema.validate({
            to: this._to,
            amount: this._amount,
            closeRemainderTo: this._closeRemainderTo,
        });
        if (validationResult.error) {
            throw new sdk_core_1.InvalidTransactionError(`Transaction validation failed: ${validationResult.error.message}`);
        }
    }
    /**
     * Sets the payment receiver.
     *
     * @param {BaseAddress} to The receiver account
     * @returns {TransferBuilder} This transfer builder.
     *
     */
    to(to) {
        this.validateAddress(to);
        this._to = to.address;
        return this;
    }
    /**
     * Sets the amount of payment.
     *
     * @param {number} amount The amount of payment.
     * @returns {TransferBuilder} This transfer builder.
     */
    amount(amount) {
        this.validateValue(new bignumber_js_1.default(amount.toString()));
        this._amount = amount;
        return this;
    }
    /**
     * Sets address to transfer remainder amount on closing.
     *
     * @param {BaseAddress} closeRemainderTo The address for receiving remainder amount
     * @returns {TransferBuilder} This transfer builder.
     *
     */
    closeRemainderTo(closeRemainderTo) {
        this.validateAddress(closeRemainderTo);
        this._closeRemainderTo = closeRemainderTo.address;
        return this;
    }
}
exports.TransferBuilder = TransferBuilder;
