// @bitgo/sdk-coin-algo/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const algosdk_1 = __importDefault(require("algosdk"));
const keyRegistrationBuilder_1 = require("./keyRegistrationBuilder");
const transferBuilder_1 = require("./transferBuilder");
const assetTransferBuilder_1 = require("./assetTransferBuilder");
const utils_1 = __importDefault(require("./utils"));
const sdk_core_1 = require("@bitgo/sdk-core");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    getTransferBuilder() {
        return new transferBuilder_1.TransferBuilder(this._coinConfig);
    }
    getAssetTransferBuilder() {
        return new assetTransferBuilder_1.AssetTransferBuilder(this._coinConfig);
    }
    getKeyRegistrationBuilder() {
        return new keyRegistrationBuilder_1.KeyRegistrationBuilder(this._coinConfig);
    }
    from(rawTxn) {
        const builder = this.getBuilder(rawTxn);
        builder.from(rawTxn);
        return builder;
    }
    getBuilder(rawTxn) {
        const decodeTxn = utils_1.default.decodeAlgoTxn(rawTxn);
        const algoTxn = decodeTxn.txn;
        if (algoTxn.type == algosdk_1.default.TransactionType.keyreg) {
            return this.getWalletInitializationBuilder();
        }
        else if (algoTxn.type == algosdk_1.default.TransactionType.pay) {
            return this.getTransferBuilder();
        }
        else if (algoTxn.type == algosdk_1.default.TransactionType.axfer) {
            return this.getAssetTransferBuilder();
        }
        else {
            throw new sdk_core_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
    }
    getWalletInitializationBuilder() {
        return new keyRegistrationBuilder_1.KeyRegistrationBuilder(this._coinConfig);
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
