// @bitgo/sdk-coin-xtz/dist/src/lib/transferBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
/** Tezos transfer builder. */
// TODO: implement BaseTransferBuilder
class TransferBuilder {
    amount(amount) {
        this._amount = amount;
        return this;
    }
    coin(coin) {
        this._coin = coin;
        return this;
    }
    from(address) {
        this._fromAddress = address;
        return this;
    }
    to(address) {
        this._toAddress = address;
        return this;
    }
    fee(fee) {
        this._fee = fee;
        return this;
    }
    gasLimit(gasLimit) {
        this._gasLimit = gasLimit;
        return this;
    }
    storageLimit(storageLimit) {
        this._storageLimit = storageLimit;
        return this;
    }
    dataToSign(dataToSign) {
        this._dataToSign = dataToSign;
        return this;
    }
    counter(counter) {
        this._counter = counter;
        return this;
    }
    build() {
        if (!this._amount || !this._fromAddress || !this._toAddress || !this._fee) {
            throw new sdk_core_1.BuildTransactionError('Missing transfer mandatory fields. Amount, from address, destination (to) address and fee are mandatory');
        }
        return {
            amount: this._amount,
            coin: this._coin,
            from: this._fromAddress,
            to: this._toAddress,
            fee: {
                fee: this._fee,
                gasLimit: this._gasLimit,
                storageLimit: this._storageLimit,
            },
            dataToSign: this._dataToSign,
            counter: this._counter,
        };
    }
}
exports.TransferBuilder = TransferBuilder;
