// @bitgo/sdk-coin-avaxp/dist/src/lib/transactionBuilderFactory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderFactory = void 0;
const avalanchejs_1 = require("@bitgo-forks/avalanchejs");
const sdk_core_1 = require("@bitgo/sdk-core");
const avalanche_1 = require("avalanche");
const evm_1 = require("avalanche/dist/apis/evm");
const platformvm_1 = require("avalanche/dist/apis/platformvm");
const exportInCTxBuilder_1 = require("./exportInCTxBuilder");
const exportTxBuilder_1 = require("./exportTxBuilder");
const importInCTxBuilder_1 = require("./importInCTxBuilder");
const importTxBuilder_1 = require("./importTxBuilder");
const permissionlessValidatorTxBuilder_1 = require("./permissionlessValidatorTxBuilder");
const utils_1 = __importDefault(require("./utils"));
const validatorTxBuilder_1 = require("./validatorTxBuilder");
class TransactionBuilderFactory extends sdk_core_1.BaseTransactionBuilderFactory {
    constructor(_coinConfig) {
        super(_coinConfig);
        this.recoverSigner = false;
    }
    /** @inheritdoc */
    from(raw) {
        var _a;
        utils_1.default.validateRawTransaction(raw);
        let txSource = 'PVM';
        let transactionBuilder = undefined;
        let tx;
        const rawNoHex = utils_1.default.removeHexPrefix(raw);
        try {
            tx = new platformvm_1.Tx();
            // could throw an error if a txType doesn't match.
            tx.fromBuffer(avalanche_1.Buffer.from(rawNoHex, 'hex'));
            if (!utils_1.default.isTransactionOf(tx, this._coinConfig.network.blockchainID)) {
                throw new Error('It is not a transaction of this platformvm old flow');
            }
        }
        catch (e) {
            try {
                txSource = 'EVM';
                tx = new evm_1.Tx();
                tx.fromBuffer(avalanche_1.Buffer.from(rawNoHex, 'hex'));
                if (!utils_1.default.isTransactionOf(tx, this._coinConfig.network.cChainBlockchainID)) {
                    throw new Error('It is not a transaction of this network or C chain EVM');
                }
            }
            catch (e) {
                try {
                    txSource = 'PVM';
                    // this should be the last because other PVM functions are still being detected in the new SDK
                    const manager = avalanchejs_1.utils.getManagerForVM('PVM');
                    const [codec, txBytes] = manager.getCodecFromBuffer(avalanchejs_1.utils.hexToBuffer(raw));
                    const unpackedTx = codec.UnpackPrefix(txBytes);
                    // A signed transaction includes 4 bytes for the number of credentials as an Int type that is not known by the codec
                    // We'll skip those 4 bytes, instead we'll loop until we've parsed all credentials
                    // @see https://docs.avax.network/reference/avalanchego/p-chain/txn-format#signed-transaction-example
                    const credentials = [];
                    let credentialBytes = unpackedTx[1].slice(4);
                    let moreCredentials = true;
                    do {
                        try {
                            const [credential, rest] = codec.UnpackPrefix(credentialBytes);
                            credentials.push(credential);
                            credentialBytes = rest;
                        }
                        catch (e) {
                            moreCredentials = false;
                        }
                    } while (credentialBytes.length > 0 && moreCredentials);
                    const unpacked = codec.UnpackPrefix(txBytes);
                    const permissionlessValidatorTx = unpacked[0];
                    const outputs = permissionlessValidatorTx.baseTx.outputs;
                    const output = outputs[0].output;
                    if (outputs[0].getAssetId() !== this._coinConfig.network.avaxAssetID) {
                        throw new Error('The Asset ID of the output does not match the transaction');
                    }
                    const fromAddresses = output.outputOwners.addrs.map((a) => avalanchejs_1.utils.hexToBuffer(a.toHex()));
                    const addressMaps = fromAddresses.map((a) => new avalanchejs_1.utils.AddressMap([[new avalanchejs_1.Address(a), 0]]));
                    tx = new avalanchejs_1.UnsignedTx(unpacked[0], [], new avalanchejs_1.utils.AddressMaps(addressMaps), credentials);
                }
                catch (e) {
                    throw new Error(`The transaction type is not recognized as an old PVM or old EVM transaction. Additionally, parsing of the new PVM AddPermissionlessValidatorTx type failed: ${e.message}`);
                }
            }
        }
        if (txSource === 'PVM') {
            if (((_a = tx === null || tx === void 0 ? void 0 : tx.tx) === null || _a === void 0 ? void 0 : _a._type) && permissionlessValidatorTxBuilder_1.PermissionlessValidatorTxBuilder.verifyTxType(tx.tx._type)) {
                transactionBuilder = this.getPermissionlessValidatorTxBuilder();
                transactionBuilder.initBuilder(tx);
            }
            else if (validatorTxBuilder_1.ValidatorTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
                transactionBuilder = this.getValidatorBuilder();
                transactionBuilder.initBuilder(tx);
            }
            else if (exportTxBuilder_1.ExportTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
                transactionBuilder = this.getExportBuilder();
                transactionBuilder.initBuilder(tx);
            }
            else if (importTxBuilder_1.ImportTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
                transactionBuilder = this.getImportBuilder();
                transactionBuilder.initBuilder(tx);
            }
        }
        else if (txSource === 'EVM') {
            if (importInCTxBuilder_1.ImportInCTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
                transactionBuilder = this.getImportInCBuilder();
                transactionBuilder.initBuilder(tx);
            }
            else if (exportInCTxBuilder_1.ExportInCTxBuilder.verifyTxType(tx.getUnsignedTx().getTransaction())) {
                transactionBuilder = this.getExportInCBuilder();
                transactionBuilder.initBuilder(tx);
            }
        }
        if (transactionBuilder === undefined) {
            throw new sdk_core_1.NotSupported('Transaction cannot be parsed or has an unsupported transaction type');
        }
        return transactionBuilder;
    }
    /** @inheritdoc */
    getTransferBuilder() {
        throw new sdk_core_1.NotSupported('Transfer is not supported in P Chain');
    }
    /**
     * Initialize Validator builder
     *
     * @returns {ValidatorTxBuilder} the builder initialized
     */
    getValidatorBuilder() {
        return new validatorTxBuilder_1.ValidatorTxBuilder(this._coinConfig);
    }
    /**
     * Initialize Permissionless Validator builder
     *
     * @returns {PermissionlessValidatorTxBuilder} the builder initialized
     */
    getPermissionlessValidatorTxBuilder() {
        return new permissionlessValidatorTxBuilder_1.PermissionlessValidatorTxBuilder(this._coinConfig);
    }
    /**
     * Export Cross chain transfer
     *
     * @returns {ExportTxBuilder} the builder initialized
     */
    getExportBuilder() {
        return new exportTxBuilder_1.ExportTxBuilder(this._coinConfig);
    }
    /**
     * Import Cross chain transfer
     *
     * @returns {ImportTxBuilder} the builder initialized
     */
    getImportBuilder() {
        return new importTxBuilder_1.ImportTxBuilder(this._coinConfig);
    }
    /**
     * Import in C chain Cross chain transfer
     *
     * @returns {ImportInCTxBuilder} the builder initialized
     */
    getImportInCBuilder() {
        return new importInCTxBuilder_1.ImportInCTxBuilder(this._coinConfig);
    }
    /**
     * Export in C chain Cross chain transfer
     *
     * @returns {ExportInCTxBuilder} the builder initialized
     */
    getExportInCBuilder() {
        return new exportInCTxBuilder_1.ExportInCTxBuilder(this._coinConfig);
    }
    /** @inheritdoc */
    getWalletInitializationBuilder() {
        throw new sdk_core_1.NotSupported('Wallet initialization is not needed');
    }
}
exports.TransactionBuilderFactory = TransactionBuilderFactory;
