// @bitgo/sdk-coin-sol/dist/src/lib/stakingAuthorizeBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakingAuthorizeBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const constants_1 = require("./constants");
const assert_1 = __importDefault(require("assert"));
const utils_1 = require("./utils");
class StakingAuthorizeBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingAuthorize;
    }
    /** @inheritdoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.StakingAuthorize) {
                const AuthorizeInstruction = instruction;
                this.stakingAddress(AuthorizeInstruction.params.stakingAddress);
                this.newAuthorizedAddress(AuthorizeInstruction.params.newAuthorizeAddress);
                this.oldAuthorizedAddress(AuthorizeInstruction.params.oldAuthorizeAddress);
            }
        }
    }
    /**
     * The address of the staking account.
     *
     * @param {string} stakingAddress public address of the staking account.
     * @returns {StakeBuilder} This staking builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    stakingAddress(stakingAddress) {
        (0, utils_1.validateAddress)(stakingAddress, 'stakingAddress');
        this._stakingAddress = stakingAddress;
        return this;
    }
    /**
     * The address of the new staking account authorization.
     *
     * @param {string} stakingAddress public address of the staking account.
     * @returns {StakeBuilder} This staking builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    newAuthorizedAddress(newAuthorizedAddress) {
        (0, utils_1.validateAddress)(newAuthorizedAddress, 'newAuthorizedAddress');
        this._newAuthorizedAddress = newAuthorizedAddress;
        return this;
    }
    /**
     * The address of the new staking account authorization.
     *
     * @param {string} stakingAddress public address of the staking account.
     * @returns {StakeBuilder} This staking builder.
     *
     * @see https://docs.solana.com/staking/stake-accounts#account-address
     */
    oldAuthorizedAddress(oldAuthorizedAddress) {
        (0, utils_1.validateAddress)(oldAuthorizedAddress, 'oldAuthorizedAddress');
        this._oldAuthorizedAddress = oldAuthorizedAddress;
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        (0, assert_1.default)(this._stakingAddress, 'Staking Address must be set before building the transaction');
        (0, assert_1.default)(this._newAuthorizedAddress, 'new authorized Address must be set before building the transaction');
        (0, assert_1.default)(this._oldAuthorizedAddress, 'old authorized Address must be set before building the transaction');
        const stakingAccountData = {
            type: constants_1.InstructionBuilderTypes.StakingAuthorize,
            params: {
                stakingAddress: this._stakingAddress,
                newWithdrawAddress: this._newAuthorizedAddress,
                newAuthorizeAddress: this._newAuthorizedAddress,
                oldAuthorizeAddress: this._oldAuthorizedAddress,
            },
        };
        this._instructionsData = [stakingAccountData];
        return await super.buildImplementation();
    }
}
exports.StakingAuthorizeBuilder = StakingAuthorizeBuilder;
