// @bitgo/sdk-coin-sol/dist/src/lib/stakingWithdrawBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StakingWithdrawBuilder = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const transactionBuilder_1 = require("./transactionBuilder");
const constants_1 = require("./constants");
const assert_1 = __importDefault(require("assert"));
const utils_1 = require("./utils");
class StakingWithdrawBuilder extends transactionBuilder_1.TransactionBuilder {
    constructor(_coinConfig) {
        super(_coinConfig);
    }
    get transactionType() {
        return sdk_core_1.TransactionType.StakingWithdraw;
    }
    /** @inheritdoc */
    initBuilder(tx) {
        super.initBuilder(tx);
        for (const instruction of this._instructionsData) {
            if (instruction.type === constants_1.InstructionBuilderTypes.StakingWithdraw) {
                const withdrawInstruction = instruction;
                this.sender(withdrawInstruction.params.fromAddress);
                this.stakingAddress(withdrawInstruction.params.stakingAddress);
                this.amount(withdrawInstruction.params.amount);
            }
        }
    }
    /**
     * The address of the staking account.
     *
     * @param {string} stakingAddress public address of the staking account
     * @returns {StakeBuilder} This staking builder.
     *
     */
    stakingAddress(stakingAddress) {
        (0, utils_1.validateAddress)(stakingAddress, 'stakingAddress');
        this._stakingAddress = stakingAddress;
        return this;
    }
    /**
     * The amount to withdraw expressed in Lamports, 1 SOL = 1_000_000_000 lamports
     *
     * @param {string} amount expressed in Lamports
     * @returns {StakeBuilder} This staking builder.
     *
     */
    amount(amount) {
        if (!(0, utils_1.isValidStakingAmount)(amount)) {
            throw new sdk_core_1.BuildTransactionError('Value cannot be zero or less');
        }
        this._amount = amount;
        return this;
    }
    /** @inheritdoc */
    async buildImplementation() {
        (0, assert_1.default)(this._sender, 'Sender must be set before building the transaction');
        (0, assert_1.default)(this._stakingAddress, 'Staking address must be set before building the transaction');
        (0, assert_1.default)(this._amount, 'Amount must be set before building the transaction');
        if (this._sender === this._stakingAddress) {
            throw new sdk_core_1.BuildTransactionError('Sender address cannot be the same as the Staking address');
        }
        const stakingWithdrawData = {
            type: constants_1.InstructionBuilderTypes.StakingWithdraw,
            params: {
                fromAddress: this._sender,
                stakingAddress: this._stakingAddress,
                amount: this._amount,
            },
        };
        this._instructionsData = [stakingWithdrawData];
        return await super.buildImplementation();
    }
}
exports.StakingWithdrawBuilder = StakingWithdrawBuilder;
