// @bitgo/sdk-coin-osmo/dist/src/osmo.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Osmo = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const statics_1 = require("@bitgo/statics");
const keyPair_1 = require("./lib/keyPair");
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const lib_1 = require("./lib");
const utils_1 = __importDefault(require("./lib/utils"));
class Osmo extends abstract_cosmos_1.CosmosCoin {
    constructor(bitgo, staticsCoin) {
        super(bitgo, staticsCoin);
        if (!staticsCoin) {
            throw new Error('missing required constructor parameter staticsCoin');
        }
        this._staticsCoin = staticsCoin;
    }
    static createInstance(bitgo, staticsCoin) {
        return new Osmo(bitgo, staticsCoin);
    }
    /** @inheritDoc **/
    getBuilder() {
        return new lib_1.TransactionBuilderFactory(statics_1.coins.get(this.getChain()));
    }
    /** @inheritDoc **/
    getBaseFactor() {
        return 1e6;
    }
    /** @inheritDoc **/
    isValidAddress(address) {
        return (utils_1.default.isValidAddress(address) || utils_1.default.isValidValidatorAddress(address) || utils_1.default.isValidContractAddress(address));
    }
    /** @inheritDoc **/
    getPublicNodeUrl() {
        return sdk_core_1.Environments[this.bitgo.getEnv()].osmoNodeUrl;
    }
    /** @inheritDoc **/
    getDenomination() {
        return statics_1.BaseUnit.OSMO;
    }
    /** @inheritDoc **/
    getGasAmountDetails() {
        return {
            gasAmount: '7000',
            gasLimit: 250000,
        };
    }
    /** @inheritDoc **/
    getKeyPair(publicKey) {
        return new keyPair_1.KeyPair({ pub: publicKey });
    }
    /** @inheritDoc **/
    getAddressFromPublicKey(publicKey) {
        return new keyPair_1.KeyPair({ pub: publicKey }).getAddress();
    }
    /** @inheritDoc **/
    valuelessTransferAllowed() {
        return true;
    }
}
exports.Osmo = Osmo;
