// @bitgo/sdk-coin-islm/dist/src/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IslmUtils = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const abstract_cosmos_1 = require("@bitgo/abstract-cosmos");
const constants = __importStar(require("./constants"));
const encoding_1 = require("@cosmjs/encoding");
const any_1 = require("cosmjs-types/google/protobuf/any");
const ethSecp256k1_1 = require("../../resources/types/ethSecp256k1");
const keccak_1 = __importDefault(require("keccak"));
class IslmUtils extends abstract_cosmos_1.CosmosUtils {
    /** @inheritdoc */
    isValidAddress(address) {
        return this.isValidCosmosLikeAddressWithMemoId(address, constants.accountAddressRegex);
    }
    /** @inheritdoc */
    isValidValidatorAddress(address) {
        return constants.validatorAddressRegex.test(address);
    }
    /** @inheritdoc */
    validateAmount(amount) {
        const amountBig = (0, bignumber_js_1.default)(amount.amount);
        if (amountBig.isLessThanOrEqualTo(0)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid amount: ' + amount.amount);
        }
        if (!constants.validDenoms.find((denom) => denom === amount.denom)) {
            throw new sdk_core_1.InvalidTransactionError('transactionBuilder: validateAmount: Invalid denom: ' + amount.denom);
        }
    }
    /** @inheritdoc */
    getPublicKeyFromDecodedTx(decodedTx) {
        var _a, _b, _c, _d;
        const publicKeyUInt8Array = (_b = (_a = decodedTx.authInfo.signerInfos) === null || _a === void 0 ? void 0 : _a[0].publicKey) === null || _b === void 0 ? void 0 : _b.value;
        if (publicKeyUInt8Array) {
            return (0, encoding_1.toHex)((0, encoding_1.fromBase64)((_d = this.decodePubkey((_c = decodedTx.authInfo.signerInfos) === null || _c === void 0 ? void 0 : _c[0].publicKey)) === null || _d === void 0 ? void 0 : _d.value));
        }
        return undefined;
    }
    /**
     * Decodes a single pubkey from ptotobuf `Any` into `Pubkey`.
     * @param {Any} pubkey
     * @returns {Pubkey} the Amino JSON representation (type/value wrapper) of the pubkey
     */
    decodePubkey(pubkey) {
        if (!pubkey || !pubkey.value) {
            return null;
        }
        const { key } = ethSecp256k1_1.PubKey.decode(pubkey.value);
        return this.encodeEthSecp256k1Pubkey(key);
    }
    /** @inheritdoc */
    getEncodedPubkey(pubkey) {
        return this.encodePubkey(this.encodeEthSecp256k1Pubkey((0, encoding_1.fromHex)(pubkey)));
    }
    /**
     * Takes a pubkey in the Amino JSON object style (type/value wrapper)
     * and convertes it into a protobuf `Any`.
     * @param {Pubkey} pubkey Amino JSON object style pubkey
     * @returns {Any} pubkey encoded as protobuf `Any`
     */
    encodePubkey(pubkey) {
        const pubkeyProto = ethSecp256k1_1.PubKey.fromPartial({
            key: (0, encoding_1.fromBase64)(pubkey.value),
        });
        return any_1.Any.fromPartial({
            typeUrl: abstract_cosmos_1.PubKeyTypeUrl.ethSecp256k1,
            value: Uint8Array.from(ethSecp256k1_1.PubKey.encode(pubkeyProto).finish()),
        });
    }
    /**
     * Takes a public key as raw bytes and returns the Amino JSON
     * representation of it (type/value wrapper).
     * @param {Uint8Array} pubkey public key as raw bytes
     * @returns {Any} Amino JSON style pubkey
     */
    encodeEthSecp256k1Pubkey(pubkey) {
        if (pubkey.length !== 33 || (pubkey[0] !== 0x02 && pubkey[0] !== 0x03)) {
            throw new Error('Public key must be compressed ethSecp256k1, i.e. 33 bytes starting with 0x02 or 0x03');
        }
        return {
            type: abstract_cosmos_1.PubKeyType.ethSecp256k1,
            value: (0, encoding_1.toBase64)(pubkey),
        };
    }
    /** @inheritdoc */
    getHashFunction() {
        return (0, keccak_1.default)('keccak256');
    }
}
exports.IslmUtils = IslmUtils;
const islmUtils = new IslmUtils();
exports.default = islmUtils;
