// @bitgo/unspents/dist/bin/generate_tables.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs/promises"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
const src_1 = require("../src");
const headers = [
    'Script Type',
    'Chain Codes',
    'Spend Type',
    'Input Size (Virtual Bytes)',
    'Relative Size (p2trMusig2 = 1.00)',
];
const relativeCostRef = src_1.Dimensions.fromScriptType('taprootKeyPathSpend').getInputsVSize();
function formatMarkdownTable(headers, rows) {
    return [headers, headers.map(() => '---'), ...rows].map((row) => `| ${row.join(' | ')} |`).join('\n');
}
function generateRowsForScriptType(headers, t, params) {
    var _a;
    const chainCode = utxolib.bitgo.toChainPair(t);
    if (!params) {
        if (t === 'p2tr') {
            return [
                ...generateRowsForScriptType(headers, t, {
                    spendTypeName: 'Script Path, Level 2 (Backup/User, Backup/BitGo)',
                    scriptTypeParams: { scriptPathLevel: 2 },
                }),
                ...generateRowsForScriptType(headers, t, {
                    spendTypeName: 'Script Path, Level 1 (User/BitGo)',
                    scriptTypeParams: { scriptPathLevel: 1 },
                }),
            ];
        }
        if (t === 'p2trMusig2') {
            return [
                ...generateRowsForScriptType(headers, t, {
                    spendTypeName: 'Script Path (Backup/User, Backup/BitGo)',
                    scriptTypeParams: { scriptPathLevel: 1 },
                }),
                ...generateRowsForScriptType(headers, t, {
                    spendTypeName: 'Key Path (User/BitGo)',
                    scriptTypeParams: { scriptPathLevel: undefined },
                }),
            ];
        }
    }
    const inputVSize = src_1.Dimensions.fromScriptType(t, params === null || params === void 0 ? void 0 : params.scriptTypeParams).getInputsVSize();
    const row = [
        t,
        chainCode.join(`/`),
        (_a = params === null || params === void 0 ? void 0 : params.spendTypeName) !== null && _a !== void 0 ? _a : 'all',
        inputVSize.toString(),
        (inputVSize / relativeCostRef).toFixed(2),
    ];
    return [row];
}
function generateTables() {
    const scriptTypes = [...utxolib.bitgo.outputScripts.scriptTypes2Of3];
    return formatMarkdownTable(headers, scriptTypes.flatMap((s) => generateRowsForScriptType(headers, s)));
}
function generateDocument() {
    return [
        '# Input Costs',
        'This document contains the worst-case input costs for various script types and spend types.',
        'The input costs are calculated using the `Dimensions` class from `@bitgo/unspents`.',
        '',
        generateTables(),
    ].join('\n');
}
if (require.main === module) {
    const outfile = 'docs/input-costs.md';
    fs.writeFile(outfile, generateDocument())
        .then(() => console.log('wrote to', outfile))
        .catch((e) => console.error(e));
}
