// @bitgo/sdk-core/dist/src/bitgo/keychain/ovcJsonCodec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BitGoKeyFromOvcShares = exports.BitGoToOvcJSON = exports.OvcToBitGoJSON = exports.OvcToOtherShare = exports.OvcShare = void 0;
/* eslint-disable no-redeclare */
const t = __importStar(require("io-ts"));
exports.OvcShare = t.intersection([
    t.type({
        publicShare: t.string,
        privateShare: t.string,
        vssProof: t.string,
        i: t.number,
        j: t.number,
    }),
    t.partial({
        paillierPublicKey: t.string,
    }),
], 'OvcShare');
exports.OvcToOtherShare = t.intersection([
    exports.OvcShare,
    t.type({
        uSig: t.string,
    }),
], 'OvcToOtherShare');
exports.OvcToBitGoJSON = t.strict({
    tssVersion: t.string,
    walletType: t.string,
    coin: t.string,
    state: t.number,
    ovc: t.type({
        1: t.type({
            gpgPubKey: t.string,
            ovcToBitgoShare: exports.OvcToOtherShare,
        }),
        2: t.type({
            gpgPubKey: t.string,
            ovcToBitgoShare: exports.OvcToOtherShare,
            ovcToOvcShare: exports.OvcToOtherShare,
        }),
    }),
}, 'OvcToBitGoJSON');
exports.BitGoToOvcJSON = t.strict({
    wallet: t.intersection([
        exports.OvcToBitGoJSON,
        t.type({
            platform: t.type({
                commonKeychain: t.string,
                walletGpgPubKeySigs: t.string,
                ovc: t.type({
                    // BitGo to User (OVC-1)
                    1: t.type({
                        bitgoToOvcShare: exports.OvcShare,
                    }),
                    // BitGo to Backup (OVC-2)
                    2: t.type({
                        bitgoToOvcShare: exports.OvcShare,
                    }),
                }),
            }),
        }),
    ]),
}, 'BitgoToOvcJson');
exports.BitGoKeyFromOvcShares = t.strict({
    bitGoOutputJsonForOvc: exports.BitGoToOvcJSON,
    bitGoKeyId: t.string,
}, 'BitGoKeyFromOvcShares');
