// @bitgo/sdk-core/dist/src/bitgo/utils/tss/ecdsa/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseEcdsaUtils = void 0;
const elliptic_1 = require("elliptic");
const baseTSSUtils_1 = __importDefault(require("../baseTSSUtils"));
const opengpgUtils_1 = require("../../opengpgUtils");
/** @inheritdoc */
class BaseEcdsaUtils extends baseTSSUtils_1.default {
    constructor(bitgo, baseCoin, wallet) {
        super(bitgo, baseCoin, wallet);
        this.setBitgoGpgPubKey(bitgo);
    }
    async setBitgoGpgPubKey(bitgo) {
        const { mpcV1, mpcV2 } = await (0, opengpgUtils_1.getBitgoGpgPubKey)(bitgo);
        this.bitgoPublicGpgKey = mpcV1;
        this.bitgoMPCv2PublicGpgKey = mpcV2;
    }
    async getBitgoPublicGpgKey() {
        if (!this.bitgoPublicGpgKey) {
            // retry getting bitgo's gpg key
            await this.setBitgoGpgPubKey(this.bitgo);
            if (!this.bitgoPublicGpgKey) {
                throw new Error("Failed to get Bitgo's gpg key");
            }
        }
        return this.bitgoPublicGpgKey;
    }
    /**
     * Gets backup pub gpg key string
     * if a third party provided then get from trust
     * @param isThirdPartyBackup
     */
    async getBackupGpgPubKey(isThirdPartyBackup = false) {
        return isThirdPartyBackup ? (0, opengpgUtils_1.getTrustGpgPubKey)(this.bitgo) : (0, opengpgUtils_1.generateGPGKeyPair)('secp256k1');
    }
    /**
     * util function that checks that a commonKeychain is valid and can ultimately resolve to a valid public key
     * @param commonKeychain - a user uploaded commonKeychain string
     * @throws if the commonKeychain is invalid length or invalid format
     */
    static validateCommonKeychainPublicKey(commonKeychain) {
        const pub = BaseEcdsaUtils.getPublicKeyFromCommonKeychain(commonKeychain);
        const secp256k1 = new elliptic_1.ec('secp256k1');
        const key = secp256k1.keyFromPublic(pub, 'hex');
        return key.getPublic().encode('hex', false).slice(2);
    }
    /**
     * Gets the common public key from commonKeychain.
     *
     * @param {String} commonKeychain common key chain between n parties
     * @returns {string} encoded public key
     */
    static getPublicKeyFromCommonKeychain(commonKeychain) {
        if (commonKeychain.length !== 130) {
            throw new Error(`Invalid commonKeychain length, expected 130, got ${commonKeychain.length}`);
        }
        const commonPubHexStr = commonKeychain.slice(0, 66);
        return commonPubHexStr;
    }
}
exports.BaseEcdsaUtils = BaseEcdsaUtils;
