// @bitgo/sdk-core/dist/src/bitgo/pendingApproval/iPendingApproval.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Type = exports.State = exports.OwnerType = void 0;
var OwnerType;
(function (OwnerType) {
    OwnerType["WALLET"] = "wallet";
    OwnerType["ENTERPRISE"] = "enterprise";
})(OwnerType = exports.OwnerType || (exports.OwnerType = {}));
var State;
(function (State) {
    State["PENDING"] = "pending";
    State["AWAITING_SIGNATURE"] = "awaitingSignature";
    State["PENDING_BITGO_ADMIN_APPROVAL"] = "pendingBitGoAdminApproval";
    State["PENDING_ID_VERIFICATION"] = "pendingIdVerification";
    State["PENDING_CUSTODIAN_APPROVAL"] = "pendingCustodianApproval";
    State["PENDING_FINAL_APPROVAL"] = "pendingFinalApproval";
    State["APPROVED"] = "approved";
    State["PROCESSING"] = "processing";
    State["REJECTED"] = "rejected";
})(State = exports.State || (exports.State = {}));
var Type;
(function (Type) {
    Type["USER_CHANGE_REQUEST"] = "userChangeRequest";
    Type["TRANSACTION_REQUEST"] = "transactionRequest";
    Type["POLICY_RULE_REQUEST"] = "policyRuleRequest";
    Type["UPDATE_APPROVALS_REQUIRED_REQUEST"] = "updateApprovalsRequiredRequest";
    Type["TRANSACTION_REQUEST_FULL"] = "transactionRequestFull";
})(Type = exports.Type || (exports.Type = {}));
