// @bitgo/sdk-coin-eos/dist/src/lib/keyPair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyPair3 = exports.KeyPair2 = exports.KeyPair1 = exports.KeyPair = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
class KeyPair {
    recordKeysFromPrivateKey(prv) {
        throw new Error('Method not implemented.');
    }
    recordKeysFromPublicKey(pub) {
        throw new Error('Method not implemented.');
    }
    getKeys() {
        throw new Error('Method not implemented.');
    }
    getAddress(format) {
        throw new Error('Method not implemented.');
    }
}
exports.KeyPair = KeyPair;
class KeyPair1 extends sdk_core_1.BlsKeyPair {
}
exports.KeyPair1 = KeyPair1;
class KeyPair2 extends sdk_core_1.Ed25519KeyPair {
    recordKeysFromPrivateKeyInProtocolFormat(prv) {
        throw new Error('Method not implemented.');
    }
    recordKeysFromPublicKeyInProtocolFormat(pub) {
        throw new Error('Method not implemented.');
    }
    getAddress(format) {
        throw new Error('Method not implemented.');
    }
    getKeys() {
        throw new Error('Method not implemented.');
    }
}
exports.KeyPair2 = KeyPair2;
class KeyPair3 extends sdk_core_1.Secp256k1ExtendedKeyPair {
}
exports.KeyPair3 = KeyPair3;
