// @bitgo/sdk-core/dist/test/unit/bitgo/utils/postWithCodec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const t = __importStar(require("io-ts"));
const superagent_1 = require("superagent");
const postWithCodec_1 = require("../../../../src/bitgo/utils/postWithCodec");
describe('postWithCodec', function () {
    function getRequest(codec, body, { useEncodedBody = true } = {}) {
        const request = (0, postWithCodec_1.postWithCodec)((0, superagent_1.agent)(), 'http://example.com', codec, body, {
            useEncodedBody,
        });
        return {
            headers: request.header,
            body: request._data,
        };
    }
    function assertRequestContains(request, body) {
        assert_1.default.deepStrictEqual(request.body, body);
    }
    const codec = t.exact(t.intersection([t.type({ foo: t.string }), t.partial({ bar: t.unknown })]));
    it('has expected values with value matching codec', function () {
        assertRequestContains(getRequest(codec, { foo: 'bar' }), { foo: 'bar' });
        assertRequestContains(getRequest(codec, { foo: 'bar', bar: null }), { foo: 'bar', bar: null });
    });
    it('has expected values with value not matching codec', function () {
        // invalid value
        assertRequestContains(getRequest(codec, { foo: null }), { foo: null });
        // non-exact value
        assertRequestContains(getRequest(codec, { foo: 'bar', boo: 1 }), { foo: 'bar' });
        // non-exact value, useEncodedBody=false
        assertRequestContains(getRequest(codec, { foo: 'bar', boo: 1 }, { useEncodedBody: false }), {
            foo: 'bar',
            boo: 1,
        });
    });
});
