// @bitgo/sdk-core/dist/src/bitgo/enterprise/enterprises.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Enterprises = void 0;
/**
 * @prettier
 */
const _ = __importStar(require("lodash"));
const enterprise_1 = require("./enterprise");
class Enterprises {
    constructor(bitgo, baseCoin) {
        this.bitgo = bitgo;
        this.baseCoin = baseCoin;
    }
    /**
     * List all enterprises available to the current user
     * @param params unused
     */
    async list(params = {}) {
        const response = (await this.bitgo.get(this.bitgo.url('/enterprise')).result());
        return response.enterprises.map((e) => {
            // instantiate a new object for each enterprise
            return new enterprise_1.Enterprise(this.bitgo, this.baseCoin, e);
        });
    }
    /**
     * Fetch an enterprise from BitGo
     * @param params
     */
    async get(params = {}) {
        const enterpriseId = params.id;
        if (_.isUndefined(enterpriseId)) {
            throw new Error('id must not be empty');
        }
        if (!_.isString(enterpriseId)) {
            throw new Error('id must be hexadecimal enterprise ID');
        }
        const enterpriseData = (await this.bitgo.get(this.bitgo.url(`/enterprise/${enterpriseId}`)).result());
        return new enterprise_1.Enterprise(this.bitgo, this.baseCoin, enterpriseData);
    }
    /**
     * Create a new enterprise
     * @param params
     */
    // TODO: (CT-686) Flesh out params object with valid enterprise creation parameters
    async create(params = {}) {
        const enterpriseData = (await this.bitgo.post(this.bitgo.url(`/enterprise`)).send(params).result());
        return new enterprise_1.Enterprise(this.bitgo, this.baseCoin, enterpriseData);
    }
    /**
     * Resign TSS configs of all enterprises belonging to the user with a
     * new user password / ecdh keychain
     */
    async resignTssConfigsForEnterprises(oldEcdhKeypair, newEcdhKeypair) {
        const enterpriseTssConfigsForUser = await this.bitgo
            .get(this.bitgo.url('/tssconfigs/ecdsa', 2))
            .query({
            createdBy: 'me',
        })
            .send()
            .result();
        const resigningPromises = enterpriseTssConfigsForUser.map(async (enterpriseTssConfig) => {
            var _a, _b, _c;
            const enterprise = new enterprise_1.Enterprise(this.bitgo, this.bitgo.coin('tbtc'), {
                id: enterpriseTssConfig.enterpriseId,
                name: '',
            });
            const enterpriseChallenge = (_a = enterpriseTssConfig.ecdsa.challenge) === null || _a === void 0 ? void 0 : _a.enterprise;
            const bitgoNitroChallenge = (_b = enterpriseTssConfig.ecdsa.challenge) === null || _b === void 0 ? void 0 : _b.bitgoNitroHsm;
            const bitgoInstChallenge = (_c = enterpriseTssConfig.ecdsa.challenge) === null || _c === void 0 ? void 0 : _c.bitgoInstitutionalHsm;
            await enterprise.resignEnterpriseChallenges(oldEcdhKeypair, newEcdhKeypair, enterpriseChallenge, bitgoInstChallenge, bitgoNitroChallenge);
        });
        await Promise.all(resigningPromises);
    }
}
exports.Enterprises = Enterprises;
