// @bitgo/sdk-coin-bch/dist/src/bch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bch = void 0;
const abstract_utxo_1 = require("@bitgo/abstract-utxo");
const utxolib = __importStar(require("@bitgo/utxo-lib"));
class Bch extends abstract_utxo_1.AbstractUtxoCoin {
    constructor(bitgo, network) {
        super(bitgo, network || utxolib.networks.bitcoincash);
    }
    static createInstance(bitgo) {
        return new Bch(bitgo);
    }
    getChain() {
        return 'bch';
    }
    getFamily() {
        return 'bch';
    }
    getFullName() {
        return 'Bitcoin Cash';
    }
    supportsBlockTarget() {
        return false;
    }
    /**
     * Canonicalize a Bitcoin Cash address for a specific version
     *
     * Starting on January 14th, 2018 Bitcoin Cash's bitcoin-abc node switched over to using cashaddr
     * encoding for all of their addresses in order to distinguish them from Bitcoin Core's.
     * https://www.bitcoinabc.org/cashaddr. We're sticking with the old base58 format because
     * migrating over to the new format will be laborious, and we want to see how the space evolves
     *
     * @param address may or may not be prefixed with the network, example bitcoincash:pppkt7q2axpsm2cajyjtu6x8fsh6ywauzgxmsru962 or pppkt7q2axpsm2cajyjtu6x8fsh6ywauzgxmsru962
     * @param version the version of the desired address, 'base58' or 'cashaddr', defaulting to 'base58'
     * @returns {*} address string
     */
    canonicalAddress(address, version = 'base58') {
        if (version === 'base58') {
            return utxolib.addressFormat.toCanonicalFormat(address, this.network);
        }
        if (version === 'cashaddr') {
            const script = utxolib.addressFormat.toOutputScriptTryFormats(address, this.network);
            return utxolib.addressFormat.fromOutputScriptWithFormat(script, version, this.network);
        }
        throw new Error(`invalid version ${version}`);
    }
}
exports.Bch = Bch;
