// @bitgo/utxo-lib/dist/src/testutil/keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultWalletKeys = exports.getDefaultCosigner = exports.getKeyName = exports.getUncompressedKeyTriple = exports.getKeyTriple = exports.getKey = void 0;
const bip32_1 = require("bip32");
const crypto = require("crypto");
const bitgo_1 = require("../bitgo");
const noble_ecc_1 = require("../noble_ecc");
const networks_1 = require("../networks");
const bip32 = (0, bip32_1.BIP32Factory)(noble_ecc_1.ecc);
function getKey(seed) {
    return bip32.fromSeed(crypto.createHash('sha256').update(seed).digest());
}
exports.getKey = getKey;
function getKeyTriple(seed) {
    return [getKey(seed + '.0'), getKey(seed + '.1'), getKey(seed + '.2')];
}
exports.getKeyTriple = getKeyTriple;
function getUncompressedKey(input) {
    // Using input for deterministic randomness
    return noble_ecc_1.ECPair.makeRandom({
        compressed: false,
        network: networks_1.networks.testnet,
        rng: () => {
            return Buffer.alloc(32, input);
        },
    });
}
function getUncompressedKeyTriple(inputs) {
    return [getUncompressedKey(inputs[0]), getUncompressedKey(inputs[1]), getUncompressedKey(inputs[2])];
}
exports.getUncompressedKeyTriple = getUncompressedKeyTriple;
function getKeyName(triple, k) {
    return ['user', 'backup', 'bitgo'][triple.indexOf(k)];
}
exports.getKeyName = getKeyName;
function getDefaultCosigner(keyset, signer) {
    const eq = (a, b) => a === b || (Buffer.isBuffer(a) && Buffer.isBuffer(b) && a.equals(b));
    const [user, backup, bitgo] = keyset;
    if (eq(signer, user)) {
        return bitgo;
    }
    if (eq(signer, backup)) {
        return bitgo;
    }
    if (eq(signer, bitgo)) {
        return user;
    }
    throw new Error(`signer not in pubkeys`);
}
exports.getDefaultCosigner = getDefaultCosigner;
function getDefaultWalletKeys() {
    return new bitgo_1.RootWalletKeys(getKeyTriple('default'));
}
exports.getDefaultWalletKeys = getDefaultWalletKeys;
