// @bitgo/sdk-coin-bsc/dist/src/bscToken.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BscToken = void 0;
const statics_1 = require("@bitgo/statics");
const abstract_eth_1 = require("@bitgo/abstract-eth");
const lib_1 = require("./lib");
class BscToken extends abstract_eth_1.EthLikeToken {
    constructor(bitgo, tokenConfig) {
        super(bitgo, tokenConfig, BscToken.coinNames);
    }
    static createTokenConstructor(config) {
        return super.createTokenConstructor(config, BscToken.coinNames);
    }
    static createTokenConstructors() {
        return super.createTokenConstructors(BscToken.coinNames);
    }
    getTransactionBuilder() {
        return new lib_1.TransactionBuilder(statics_1.coins.get(this.getBaseChain()));
    }
    /** @inheritDoc */
    supportsTss() {
        return true;
    }
    /** @inheritDoc */
    getMPCAlgorithm() {
        return 'ecdsa';
    }
    getFullName() {
        return 'Bsc Token';
    }
}
exports.BscToken = BscToken;
BscToken.coinNames = {
    Mainnet: 'bsc',
    Testnet: 'tbsc',
};
