// @bitgo/sdk-coin-sui/dist/src/lib/iface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MethodNames = exports.SuiTransactionType = void 0;
var SuiTransactionType;
(function (SuiTransactionType) {
    SuiTransactionType["Transfer"] = "Transfer";
    SuiTransactionType["AddStake"] = "AddStake";
    SuiTransactionType["WithdrawStake"] = "WithdrawStake";
    SuiTransactionType["CustomTx"] = "CustomTx";
})(SuiTransactionType = exports.SuiTransactionType || (exports.SuiTransactionType = {}));
/**
 * Method names for the transaction method. Names change based on the type of transaction e.g 'request_add_delegation_mul_coin' for the staking transaction
 */
var MethodNames;
(function (MethodNames) {
    /**
     * Add stake to a validator's staking pool.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/sui_system.md#function-request_add_stake
     */
    MethodNames["RequestAddStake"] = "request_add_stake";
    /**
     * Withdraw some portion of a stake from a validator's staking pool.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/sui_system.md#function-request_withdraw_stake
     */
    MethodNames["RequestWithdrawStake"] = "request_withdraw_stake";
    /**
     * Split StakedSui self to two parts, one with principal split_amount, and the remaining principal is left in self.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/staking_pool.md#0x3_staking_pool_split
     */
    MethodNames["StakingPoolSplit"] = "::staking_pool::split";
    /**
     * Transfer ownership of obj to recipient.
     *
     * @see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/docs/transfer.md#function-public_transfer
     */
    MethodNames["PublicTransfer"] = "::transfer::public_transfer";
})(MethodNames = exports.MethodNames || (exports.MethodNames = {}));
