// @bitgo/sdk-lib-mpc/dist/src/hashCommitment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyCommitment = exports.createCommitment = void 0;
const crypto_1 = require("crypto");
const util_1 = require("./util");
const minRandomnessLength = 32;
/**
 * Create hash commitment and decommietment of a secret value.
 * @param secret The secret value/message.
 * @param r The randomness/nonce to be added to the commmitment.
 * @returns The created commitment and decommitment.
 */
function createCommitment(secret, r = (0, crypto_1.randomBytes)(minRandomnessLength)) {
    if (r.length < minRandomnessLength) {
        throw new Error(`randomness must be at least ${minRandomnessLength} bytes long`);
    }
    return {
        commitment: hash(secret, r),
        decommitment: {
            blindingFactor: r,
            secret: secret,
        },
    };
}
exports.createCommitment = createCommitment;
const bytesPerUint32 = 4;
function hash(secret, r) {
    return (0, crypto_1.createHash)('sha256')
        .update((0, util_1.bigIntToBufferBE)(BigInt(secret.length), bytesPerUint32))
        .update(secret)
        .update((0, util_1.bigIntToBufferBE)(BigInt(r.length), bytesPerUint32))
        .update(r)
        .digest();
}
/**
 * Verify hash commitment and decommietment of a secret value.
 * @param commitment The commitment.
 * @param decommietment The decommitment.
 * @returns True if verification succeeds.
 */
function verifyCommitment(commitment, decommietment) {
    return hash(decommietment.secret, decommietment.blindingFactor).compare(commitment) === 0;
}
exports.verifyCommitment = verifyCommitment;
